{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7755b941",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "# Read the dataset directly\n",
    "file_path = 'df.csv'\n",
    "df = pd.read_csv(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22b6f6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_summary_df(df):\n",
    "    count_df1 = df.groupby(['transcript_id', 'transcript_position']).size().reset_index(name='read_count')\n",
    "    count_df2 = df.groupby(['transcript_id']).size().reset_index(name='expression_count') #most genes are likely to produce unique transcripts\n",
    "    df = df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "    df = df.merge(count_df2, on=['transcript_id'])\n",
    "    \n",
    "    summary_df = df.groupby(['transcript_id', 'transcript_position']).agg({\n",
    "        'gene_id': 'first',\n",
    "        'combined nucleotides': 'first',\n",
    "        'dwelling_time1': ['mean'],\n",
    "        'sd1': ['mean'],\n",
    "        'mean1': ['mean'],\n",
    "        'dwelling_time2': ['mean'],\n",
    "        'sd2': ['mean'],\n",
    "        'mean2': ['mean'],\n",
    "        'dwelling_time3': ['mean'],\n",
    "        'sd3': ['mean'],\n",
    "        'mean3': ['mean'],\n",
    "        'label': lambda x: x.mode()[0] if not x.mode().empty else None\n",
    "    }).reset_index()\n",
    "    summary_df.columns = summary_df.columns.map('_'.join)\n",
    "    summary_df = summary_df.rename(columns={'label_<lambda>': 'label',\n",
    "                                           'transcript_id_': 'transcript_id',\n",
    "                                           'transcript_position_': 'transcript_position',\n",
    "                                           'gene_id_first': 'gene_id',\n",
    "                                           'combined nucleotides_first': 'combined nucleotides'})\n",
    "    \n",
    "    return summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d944a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df = get_summary_df(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3249b1a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1_mean</th>\n",
       "      <th>sd1_mean</th>\n",
       "      <th>mean1_mean</th>\n",
       "      <th>dwelling_time2_mean</th>\n",
       "      <th>sd2_mean</th>\n",
       "      <th>mean2_mean</th>\n",
       "      <th>dwelling_time3_mean</th>\n",
       "      <th>sd3_mean</th>\n",
       "      <th>mean3_mean</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>4.005616</td>\n",
       "      <td>82.004110</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>3.644638</td>\n",
       "      <td>80.497101</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>2.181563</td>\n",
       "      <td>84.190625</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>2.540877</td>\n",
       "      <td>82.289474</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>2.576731</td>\n",
       "      <td>78.536538</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_position          gene_id  \\\n",
       "0       ENST00000000233                  244  ENSG00000004059   \n",
       "1       ENST00000000233                  261  ENSG00000004059   \n",
       "2       ENST00000000233                  316  ENSG00000004059   \n",
       "3       ENST00000000233                  332  ENSG00000004059   \n",
       "4       ENST00000000233                  368  ENSG00000004059   \n",
       "...                 ...                  ...              ...   \n",
       "121833  ENST00000641834                 1348  ENSG00000167747   \n",
       "121834  ENST00000641834                 1429  ENSG00000167747   \n",
       "121835  ENST00000641834                 1531  ENSG00000167747   \n",
       "121836  ENST00000641834                 1537  ENSG00000167747   \n",
       "121837  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "       combined nucleotides  dwelling_time1_mean  sd1_mean  mean1_mean  \\\n",
       "0                   AAGACCA             0.008264  4.223784  123.702703   \n",
       "1                   CAAACTG             0.006609  3.216424  109.681395   \n",
       "2                   GAAACAG             0.007570  2.940541  105.475676   \n",
       "3                   AGAACAT             0.010620  6.476350  129.355000   \n",
       "4                   AGGACAA             0.010701  6.415051  117.924242   \n",
       "...                     ...                  ...       ...         ...   \n",
       "121833              GGGACAT             0.009594  3.294164  118.232877   \n",
       "121834              CTGACAC             0.008393  4.511014  110.969565   \n",
       "121835              TGGACAC             0.008161  3.918438  113.968750   \n",
       "121836              CTGACCA             0.008044  3.191228  109.354386   \n",
       "121837              TTGACAT             0.008788  4.090577  105.807692   \n",
       "\n",
       "        dwelling_time2_mean  sd2_mean  mean2_mean  dwelling_time3_mean  \\\n",
       "0                  0.009373  7.382162  125.913514             0.007345   \n",
       "1                  0.006813  3.226535  107.889535             0.007710   \n",
       "2                  0.007416  3.642703   98.947027             0.007555   \n",
       "3                  0.008632  2.899200   97.836500             0.006101   \n",
       "4                  0.011479  5.870303  121.954545             0.010019   \n",
       "...                     ...       ...         ...                  ...   \n",
       "121833             0.007300  4.929726  116.342466             0.006555   \n",
       "121834             0.010305  9.105797  114.927536             0.005568   \n",
       "121835             0.006877  4.759688  113.562500             0.006410   \n",
       "121836             0.007419  6.552982  123.263158             0.006472   \n",
       "121837             0.006908  8.702885  113.134615             0.008337   \n",
       "\n",
       "        sd3_mean  mean3_mean  label  \n",
       "0       4.386989   80.570270      0  \n",
       "1       3.016599   94.290698      0  \n",
       "2       2.087146   89.364324      0  \n",
       "3       2.236520   89.154000      0  \n",
       "4       4.260253   85.178788      0  \n",
       "...          ...         ...    ...  \n",
       "121833  4.005616   82.004110      1  \n",
       "121834  3.644638   80.497101      0  \n",
       "121835  2.181563   84.190625      1  \n",
       "121836  2.540877   82.289474      0  \n",
       "121837  2.576731   78.536538      0  \n",
       "\n",
       "[121838 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63faec17",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = summary_df.drop(['label','transcript_id','combined nucleotides','gene_id'], axis=1)\n",
    "y = summary_df['label']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edb1b66",
   "metadata": {},
   "source": [
    "### Logistic Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3697096d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cross-Validation ROC AUC Scores: [0.68612394 0.66941754 0.66690584 0.70394234 0.64875752]\n",
      "Mean ROC AUC Score: 0.6750294348682209\n",
      "Standard Deviation of ROC AUC Scores: 0.018690875307401118\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Cross-Validation PR AUC Scores: [0.04493598 0.04493598 0.04493598 0.04493783 0.04493783]\n",
      "Mean PR AUC Score: 0.04493671926831497\n",
      "Standard Deviation of PR AUC Scores: 9.034368288930283e-07\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Average of ROC AUC and PR AUC Scores for Each Fold: [0.36552996 0.35717676 0.35592091 0.37444008 0.34684767]\n",
      "Mean of Average Scores: 0.359983077068268\n",
      "Standard Deviation of Average Scores: 0.009345463721837399\n"
     ]
    }
   ],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.model_selection import cross_validate\n",
    "from sklearn.metrics import make_scorer, roc_auc_score, average_precision_score\n",
    "\n",
    "# Create LogisticRegression instance\n",
    "log_reg_model = LogisticRegression(random_state=42, max_iter=1000)\n",
    "\n",
    "# Define scorers\n",
    "scorers = {\n",
    "    'roc_auc': 'roc_auc',  # Uses the ROC AUC scoring\n",
    "    'pr_auc': make_scorer(average_precision_score)  # Uses the Precision-Recall AUC scoring\n",
    "}\n",
    "\n",
    "# Cross-validation with multiple scoring metrics\n",
    "cv_results = cross_validate(log_reg_model, X, y, cv=5, scoring=scorers)\n",
    "\n",
    "# Calculate averages of ROC AUC and PR AUC for each fold\n",
    "average_scores = (cv_results['test_roc_auc'] + cv_results['test_pr_auc']) / 2\n",
    "\n",
    "print(\"Cross-Validation ROC AUC Scores:\", cv_results['test_roc_auc'])\n",
    "print(\"Mean ROC AUC Score:\", cv_results['test_roc_auc'].mean())\n",
    "print(\"Standard Deviation of ROC AUC Scores:\", cv_results['test_roc_auc'].std())\n",
    "print('-' * 100)\n",
    "print(\"Cross-Validation PR AUC Scores:\", cv_results['test_pr_auc'])\n",
    "print(\"Mean PR AUC Score:\", cv_results['test_pr_auc'].mean())\n",
    "print(\"Standard Deviation of PR AUC Scores:\", cv_results['test_pr_auc'].std())\n",
    "print('-' * 100)\n",
    "print(\"Average of ROC AUC and PR AUC Scores for Each Fold:\", average_scores)\n",
    "print(\"Mean of Average Scores:\", average_scores.mean())\n",
    "print(\"Standard Deviation of Average Scores:\", average_scores.std())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11aca7ef",
   "metadata": {},
   "source": [
    "### XGBoost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "daa36ca2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cross-Validation ROC AUC Scores: [0.89004002 0.87733013 0.87966446 0.88480393 0.88732616]\n",
      "Mean ROC AUC Score: 0.8838329401047893\n",
      "Standard Deviation of ROC AUC Scores: 0.004718812695889388\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Cross-Validation PR AUC Scores: [0.22111677 0.21106019 0.19523733 0.20722406 0.1636434 ]\n",
      "Mean PR AUC Score: 0.19965635108707397\n",
      "Standard Deviation of PR AUC Scores: 0.019820837290330188\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Average of ROC AUC and PR AUC Scores for Each Fold: [0.5555784  0.54419516 0.5374509  0.546014   0.52548478]\n",
      "Mean of Average Scores: 0.5417446455959316\n",
      "Standard Deviation of Average Scores: 0.009984233545896618\n"
     ]
    }
   ],
   "source": [
    "import xgboost as xgb\n",
    "from sklearn.model_selection import cross_validate\n",
    "from sklearn.metrics import make_scorer, roc_auc_score, average_precision_score\n",
    "\n",
    "# Create XGBClassifier instance\n",
    "xgb_model = xgb.XGBClassifier(n_estimators=100, random_state=42)\n",
    "\n",
    "# Define scorers\n",
    "scorers = {\n",
    "    'roc_auc': 'roc_auc',  # Uses the ROC AUC scoring\n",
    "    'pr_auc': make_scorer(average_precision_score)  # Uses the Precision-Recall AUC scoring\n",
    "}\n",
    "\n",
    "# Cross-validation with multiple scoring metrics\n",
    "cv_results = cross_validate(xgb_model, X, y, cv=5, scoring=scorers)\n",
    "\n",
    "# Calculate averages of ROC AUC and PR AUC for each fold\n",
    "average_scores = (cv_results['test_roc_auc'] + cv_results['test_pr_auc']) / 2\n",
    "\n",
    "print(\"Cross-Validation ROC AUC Scores:\", cv_results['test_roc_auc'])\n",
    "print(\"Mean ROC AUC Score:\", cv_results['test_roc_auc'].mean())\n",
    "print(\"Standard Deviation of ROC AUC Scores:\", cv_results['test_roc_auc'].std())\n",
    "print('-' * 100)\n",
    "print(\"Cross-Validation PR AUC Scores:\", cv_results['test_pr_auc'])\n",
    "print(\"Mean PR AUC Score:\", cv_results['test_pr_auc'].mean())\n",
    "print(\"Standard Deviation of PR AUC Scores:\", cv_results['test_pr_auc'].std())\n",
    "print('-' * 100)\n",
    "print(\"Average of ROC AUC and PR AUC Scores for Each Fold:\", average_scores)\n",
    "print(\"Mean of Average Scores:\", average_scores.mean())\n",
    "print(\"Standard Deviation of Average Scores:\", average_scores.std())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21567e58",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
