{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a51bbfde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import json\n",
    "import pandas as pd\n",
    "import xgboost as xgb\n",
    "# Read data.info table ====================================================================================================\n",
    "data = pd.read_csv(\"data.info\")\n",
    "data = pd.DataFrame(data)\n",
    "data['transcript_position'] = data['transcript_position'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29bfd038",
   "metadata": {},
   "outputs": [],
   "source": [
    "objs0=[]\n",
    "with gzip.open(\"dataset0.json.gz\", 'rt') as f:\n",
    "    for line in f:\n",
    "        objs0.append(json.loads(line))\n",
    "\n",
    "\n",
    "\n",
    "rows0 = []\n",
    "for i in range(len(objs0)):\n",
    "    for key1, inner_dict in objs0[i].items():\n",
    "        for key2, inner_list in inner_dict.items():\n",
    "            for key3, sublist in inner_list.items():\n",
    "                for subsublist in sublist:\n",
    "                    rows0.append([key1, key2, key3] + subsublist)\n",
    "                    \n",
    "df0 = pd.DataFrame(rows0, columns=[\"transcript_id\", \"transcript_position\", \"combined nucleotides\", \"dwelling_time1\", \"sd1\", \"mean1\", \"dwelling_time2\", \"sd2\", \"mean2\", \"dwelling_time3\", \"sd3\", \"mean3\"])\n",
    "df0['transcript_position'] = df0['transcript_position'].astype(int)\n",
    "\n",
    "\n",
    "#df0info = pd.merge(df0, data, on=['transcript_id', 'transcript_position'], how='inner')\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e99fa466",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0.to_pickle('df0.pkl')\n",
    "df1.to_pickle('df1.pkl')\n",
    "df2.to_pickle('df2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2666b999",
   "metadata": {},
   "outputs": [],
   "source": [
    "objs1=[]\n",
    "with gzip.open(\"dataset1.json.gz\", 'rt') as f:\n",
    "    for line in f:\n",
    "        objs1.append(json.loads(line))\n",
    "        \n",
    "rows1 = []\n",
    "for i in range(len(objs1)):\n",
    "    for key1, inner_dict in objs1[i].items():\n",
    "        for key2, inner_list in inner_dict.items():\n",
    "            for key3, sublist in inner_list.items():\n",
    "                for subsublist in sublist:\n",
    "                    rows1.append([key1, key2, key3] + subsublist)                    \n",
    "df1 = pd.DataFrame(rows1, columns=[\"transcript_id\", \"transcript_position\", \"combined nucleotides\", \"dwelling_time1\", \"sd1\", \"mean1\", \"dwelling_time2\", \"sd2\", \"mean2\", \"dwelling_time3\", \"sd3\", \"mean3\"])\n",
    "df1['transcript_position'] = df1['transcript_position'].astype(int)\n",
    "#df1info = pd.merge(df1, data, on=['transcript_id', 'transcript_position'], how='inner')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4029856a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0.to_csv('df0.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "838ad73c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11027101</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.00418</td>\n",
       "      <td>7.49</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>10.20</td>\n",
       "      <td>116.0</td>\n",
       "      <td>0.01000</td>\n",
       "      <td>2.01</td>\n",
       "      <td>76.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11027102</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>1.91</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00598</td>\n",
       "      <td>12.30</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.01760</td>\n",
       "      <td>2.61</td>\n",
       "      <td>74.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11027103</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.00721</td>\n",
       "      <td>4.58</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>6.58</td>\n",
       "      <td>113.0</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>2.28</td>\n",
       "      <td>85.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11027104</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>2.33</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00913</td>\n",
       "      <td>10.40</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>4.44</td>\n",
       "      <td>76.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11027105</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>3.13</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00303</td>\n",
       "      <td>9.98</td>\n",
       "      <td>118.0</td>\n",
       "      <td>0.01930</td>\n",
       "      <td>1.79</td>\n",
       "      <td>76.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11027106 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id  transcript_position combined nucleotides  \\\n",
       "0         ENST00000000233                  244              AAGACCA   \n",
       "1         ENST00000000233                  244              AAGACCA   \n",
       "2         ENST00000000233                  244              AAGACCA   \n",
       "3         ENST00000000233                  244              AAGACCA   \n",
       "4         ENST00000000233                  244              AAGACCA   \n",
       "...                   ...                  ...                  ...   \n",
       "11027101  ENST00000641834                 1693              TTGACAT   \n",
       "11027102  ENST00000641834                 1693              TTGACAT   \n",
       "11027103  ENST00000641834                 1693              TTGACAT   \n",
       "11027104  ENST00000641834                 1693              TTGACAT   \n",
       "11027105  ENST00000641834                 1693              TTGACAT   \n",
       "\n",
       "          dwelling_time1   sd1  mean1  dwelling_time2    sd2  mean2  \\\n",
       "0                0.00299  2.06  125.0         0.01770  10.40  122.0   \n",
       "1                0.00631  2.53  125.0         0.00844   4.67  126.0   \n",
       "2                0.00465  3.92  109.0         0.01360  12.00  124.0   \n",
       "3                0.00398  2.06  125.0         0.00830   5.01  130.0   \n",
       "4                0.00664  2.92  120.0         0.00266   3.94  129.0   \n",
       "...                  ...   ...    ...             ...    ...    ...   \n",
       "11027101         0.00418  7.49  108.0         0.00564  10.20  116.0   \n",
       "11027102         0.00664  1.91  109.0         0.00598  12.30  110.0   \n",
       "11027103         0.00721  4.58  105.0         0.00398   6.58  113.0   \n",
       "11027104         0.00266  2.33  109.0         0.00913  10.40  108.0   \n",
       "11027105         0.00564  3.13  110.0         0.00303   9.98  118.0   \n",
       "\n",
       "          dwelling_time3    sd3  mean3  \n",
       "0                0.00930  10.90   84.1  \n",
       "1                0.01030   6.30   80.9  \n",
       "2                0.00498   2.13   79.6  \n",
       "3                0.00498   3.78   80.4  \n",
       "4                0.01300   7.15   82.2  \n",
       "...                  ...    ...    ...  \n",
       "11027101         0.01000   2.01   76.4  \n",
       "11027102         0.01760   2.61   74.6  \n",
       "11027103         0.00316   2.28   85.3  \n",
       "11027104         0.00664   4.44   76.8  \n",
       "11027105         0.01930   1.79   76.2  \n",
       "\n",
       "[11027106 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "967617ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df0info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b1848f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv('df1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "90185be5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>2.16</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00640</td>\n",
       "      <td>3.90</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>8.75</td>\n",
       "      <td>83.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>4.43</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.01860</td>\n",
       "      <td>10.00</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>6.20</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>3.10</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>8.26</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01590</td>\n",
       "      <td>2.89</td>\n",
       "      <td>78.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>4.52</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.01750</td>\n",
       "      <td>8.51</td>\n",
       "      <td>128.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.63</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>2.81</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00772</td>\n",
       "      <td>4.22</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.00474</td>\n",
       "      <td>5.84</td>\n",
       "      <td>80.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7907947</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>1.20</td>\n",
       "      <td>93.5</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.58</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01530</td>\n",
       "      <td>2.19</td>\n",
       "      <td>92.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7907948</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>2.00</td>\n",
       "      <td>92.1</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>3.00</td>\n",
       "      <td>99.1</td>\n",
       "      <td>0.00469</td>\n",
       "      <td>1.37</td>\n",
       "      <td>92.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7907949</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00430</td>\n",
       "      <td>1.32</td>\n",
       "      <td>91.8</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.33</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.00550</td>\n",
       "      <td>1.37</td>\n",
       "      <td>93.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7907950</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00280</td>\n",
       "      <td>2.70</td>\n",
       "      <td>90.7</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>1.45</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.01600</td>\n",
       "      <td>1.90</td>\n",
       "      <td>92.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7907951</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>3.17</td>\n",
       "      <td>91.7</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>2.10</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01320</td>\n",
       "      <td>1.72</td>\n",
       "      <td>93.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7907952 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id  transcript_position combined nucleotides  \\\n",
       "0        ENST00000000233                  244              AAGACCA   \n",
       "1        ENST00000000233                  244              AAGACCA   \n",
       "2        ENST00000000233                  244              AAGACCA   \n",
       "3        ENST00000000233                  244              AAGACCA   \n",
       "4        ENST00000000233                  244              AAGACCA   \n",
       "...                  ...                  ...                  ...   \n",
       "7907947  ENST00000641784                 3266              CTAACTT   \n",
       "7907948  ENST00000641784                 3266              CTAACTT   \n",
       "7907949  ENST00000641784                 3266              CTAACTT   \n",
       "7907950  ENST00000641784                 3266              CTAACTT   \n",
       "7907951  ENST00000641784                 3266              CTAACTT   \n",
       "\n",
       "         dwelling_time1   sd1  mean1  dwelling_time2    sd2  mean2  \\\n",
       "0               0.00465  2.16  127.0         0.00640   3.90  127.0   \n",
       "1               0.02690  4.43  106.0         0.01860  10.00  123.0   \n",
       "2               0.00432  3.10  108.0         0.01200   8.26  125.0   \n",
       "3               0.00996  4.52  123.0         0.01750   8.51  128.0   \n",
       "4               0.00764  2.81  124.0         0.00772   4.22  126.0   \n",
       "...                 ...   ...    ...             ...    ...    ...   \n",
       "7907947         0.00589  1.20   93.5         0.00270   1.58  102.0   \n",
       "7907948         0.00401  2.00   92.1         0.00764   3.00   99.1   \n",
       "7907949         0.00430  1.32   91.8         0.00270   1.33  100.0   \n",
       "7907950         0.00280  2.70   90.7         0.00564   1.45  100.0   \n",
       "7907951         0.00332  3.17   91.7         0.00432   2.10  102.0   \n",
       "\n",
       "         dwelling_time3   sd3  mean3  \n",
       "0               0.00797  8.75   83.7  \n",
       "1               0.00863  6.20   80.0  \n",
       "2               0.01590  2.89   78.7  \n",
       "3               0.00498  2.63   80.0  \n",
       "4               0.00474  5.84   80.9  \n",
       "...                 ...   ...    ...  \n",
       "7907947         0.01530  2.19   92.3  \n",
       "7907948         0.00469  1.37   92.6  \n",
       "7907949         0.00550  1.37   93.9  \n",
       "7907950         0.01600  1.90   92.3  \n",
       "7907951         0.01320  1.72   93.5  \n",
       "\n",
       "[7907952 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793a54a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df1info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c98654ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "objs2=[]\n",
    "with gzip.open(\"dataset2.json.gz\", 'rt') as f:\n",
    "    for line in f:\n",
    "        objs2.append(json.loads(line))\n",
    "\n",
    "rows2 = []\n",
    "for i in range(len(objs2)):\n",
    "    for key1, inner_dict in objs2[i].items():\n",
    "        for key2, inner_list in inner_dict.items():\n",
    "            for key3, sublist in inner_list.items():\n",
    "                for subsublist in sublist:\n",
    "                    rows2.append([key1, key2, key3] + subsublist)     \n",
    "\n",
    "df2 = pd.DataFrame(rows2, columns=[\"transcript_id\", \"transcript_position\", \"combined nucleotides\", \"dwelling_time1\", \"sd1\", \"mean1\", \"dwelling_time2\", \"sd2\", \"mean2\", \"dwelling_time3\", \"sd3\", \"mean3\"])\n",
    "df2['transcript_position'] = df2['transcript_position'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0629c1a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv('df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "384c20e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.01220</td>\n",
       "      <td>3.99</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00337</td>\n",
       "      <td>4.56</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>4.200</td>\n",
       "      <td>84.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.03020</td>\n",
       "      <td>2.32</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.00443</td>\n",
       "      <td>2.36</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>2.130</td>\n",
       "      <td>79.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>5.55</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>7.04</td>\n",
       "      <td>99.3</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>2.210</td>\n",
       "      <td>86.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>2.10</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>3.90</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>2.180</td>\n",
       "      <td>82.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.02110</td>\n",
       "      <td>3.49</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>3.80</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.00997</td>\n",
       "      <td>2.180</td>\n",
       "      <td>81.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171935</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00675</td>\n",
       "      <td>8.33</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00759</td>\n",
       "      <td>6.15</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>2.420</td>\n",
       "      <td>91.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171936</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00529</td>\n",
       "      <td>4.81</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>7.42</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>1.940</td>\n",
       "      <td>90.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171937</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.01460</td>\n",
       "      <td>7.72</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>1.21</td>\n",
       "      <td>92.4</td>\n",
       "      <td>0.02150</td>\n",
       "      <td>2.110</td>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171938</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00610</td>\n",
       "      <td>4.21</td>\n",
       "      <td>119.0</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>5.44</td>\n",
       "      <td>94.8</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>2.000</td>\n",
       "      <td>82.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171939</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>3.02</td>\n",
       "      <td>96.8</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>1.20</td>\n",
       "      <td>91.5</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>0.913</td>\n",
       "      <td>89.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1171940 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position combined nucleotides  \\\n",
       "0             tx_id_0                    0              AAAACCT   \n",
       "1             tx_id_0                    0              AAAACCT   \n",
       "2             tx_id_0                    0              AAAACCT   \n",
       "3             tx_id_0                    0              AAAACCT   \n",
       "4             tx_id_0                    0              AAAACCT   \n",
       "...               ...                  ...                  ...   \n",
       "1171935       tx_id_6                 1880              CGAACAA   \n",
       "1171936       tx_id_6                 1880              CGAACAA   \n",
       "1171937       tx_id_6                 1880              CGAACAA   \n",
       "1171938       tx_id_6                 1880              CGAACAA   \n",
       "1171939       tx_id_6                 1880              CGAACAA   \n",
       "\n",
       "         dwelling_time1   sd1  mean1  dwelling_time2   sd2  mean2  \\\n",
       "0               0.01220  3.99  106.0         0.00337  4.56  102.0   \n",
       "1               0.03020  2.32  107.0         0.00443  2.36  102.0   \n",
       "2               0.00232  5.55  110.0         0.00664  7.04   99.3   \n",
       "3               0.00465  2.10  104.0         0.00996  3.90  108.0   \n",
       "4               0.02110  3.49  103.0         0.00531  3.80  101.0   \n",
       "...                 ...   ...    ...             ...   ...    ...   \n",
       "1171935         0.00675  8.33  111.0         0.00759  6.15  104.0   \n",
       "1171936         0.00529  4.81  106.0         0.00332  7.42  101.0   \n",
       "1171937         0.01460  7.72  104.0         0.00266  1.21   92.4   \n",
       "1171938         0.00610  4.21  119.0         0.00232  5.44   94.8   \n",
       "1171939         0.00232  3.02   96.8         0.00465  1.20   91.5   \n",
       "\n",
       "         dwelling_time3    sd3  mean3  \n",
       "0               0.00664  4.200   84.2  \n",
       "1               0.00332  2.130   79.2  \n",
       "2               0.00232  2.210   86.6  \n",
       "3               0.00401  2.180   82.2  \n",
       "4               0.00997  2.180   81.2  \n",
       "...                 ...    ...    ...  \n",
       "1171935         0.00896  2.420   91.6  \n",
       "1171936         0.00996  1.940   90.9  \n",
       "1171937         0.02150  2.110   90.0  \n",
       "1171938         0.00199  2.000   82.6  \n",
       "1171939         0.00398  0.913   89.8  \n",
       "\n",
       "[1171940 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "48aec2b2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id    transcript_id  transcript_position  label\n",
       "0       ENSG00000004059  ENST00000000233                  244      0\n",
       "1       ENSG00000004059  ENST00000000233                  261      0\n",
       "2       ENSG00000004059  ENST00000000233                  316      0\n",
       "3       ENSG00000004059  ENST00000000233                  332      0\n",
       "4       ENSG00000004059  ENST00000000233                  368      0\n",
       "...                 ...              ...                  ...    ...\n",
       "121833  ENSG00000167747  ENST00000641834                 1348      1\n",
       "121834  ENSG00000167747  ENST00000641834                 1429      0\n",
       "121835  ENSG00000167747  ENST00000641834                 1531      1\n",
       "121836  ENSG00000167747  ENST00000641834                 1537      0\n",
       "121837  ENSG00000167747  ENST00000641834                 1693      0\n",
       "\n",
       "[121838 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "45044ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0 = pd.read_pickle(\"df0.pkl\")\n",
    "df1 = pd.read_pickle(\"df1.pkl\")\n",
    "df2 = pd.read_pickle(\"df2.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9c9fd6ae",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "import xgboost as xgb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "a0c03317",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df1info = pd.merge(df1, data, on=['transcript_id', 'transcript_position'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "50e529fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>2.16</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00640</td>\n",
       "      <td>3.90</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>8.75</td>\n",
       "      <td>83.7</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>4.43</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.01860</td>\n",
       "      <td>10.00</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>6.20</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>3.10</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>8.26</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01590</td>\n",
       "      <td>2.89</td>\n",
       "      <td>78.7</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>4.52</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.01750</td>\n",
       "      <td>8.51</td>\n",
       "      <td>128.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.63</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>2.81</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00772</td>\n",
       "      <td>4.22</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.00474</td>\n",
       "      <td>5.84</td>\n",
       "      <td>80.9</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102083</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>1.20</td>\n",
       "      <td>93.5</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.58</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01530</td>\n",
       "      <td>2.19</td>\n",
       "      <td>92.3</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102084</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>2.00</td>\n",
       "      <td>92.1</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>3.00</td>\n",
       "      <td>99.1</td>\n",
       "      <td>0.00469</td>\n",
       "      <td>1.37</td>\n",
       "      <td>92.6</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102085</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00430</td>\n",
       "      <td>1.32</td>\n",
       "      <td>91.8</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.33</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.00550</td>\n",
       "      <td>1.37</td>\n",
       "      <td>93.9</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102086</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00280</td>\n",
       "      <td>2.70</td>\n",
       "      <td>90.7</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>1.45</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.01600</td>\n",
       "      <td>1.90</td>\n",
       "      <td>92.3</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102087</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>3.17</td>\n",
       "      <td>91.7</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>2.10</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01320</td>\n",
       "      <td>1.72</td>\n",
       "      <td>93.5</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7102088 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id  transcript_position combined nucleotides  \\\n",
       "0        ENST00000000233                  244              AAGACCA   \n",
       "1        ENST00000000233                  244              AAGACCA   \n",
       "2        ENST00000000233                  244              AAGACCA   \n",
       "3        ENST00000000233                  244              AAGACCA   \n",
       "4        ENST00000000233                  244              AAGACCA   \n",
       "...                  ...                  ...                  ...   \n",
       "7102083  ENST00000641784                 3266              CTAACTT   \n",
       "7102084  ENST00000641784                 3266              CTAACTT   \n",
       "7102085  ENST00000641784                 3266              CTAACTT   \n",
       "7102086  ENST00000641784                 3266              CTAACTT   \n",
       "7102087  ENST00000641784                 3266              CTAACTT   \n",
       "\n",
       "         dwelling_time1   sd1  mean1  dwelling_time2    sd2  mean2  \\\n",
       "0               0.00465  2.16  127.0         0.00640   3.90  127.0   \n",
       "1               0.02690  4.43  106.0         0.01860  10.00  123.0   \n",
       "2               0.00432  3.10  108.0         0.01200   8.26  125.0   \n",
       "3               0.00996  4.52  123.0         0.01750   8.51  128.0   \n",
       "4               0.00764  2.81  124.0         0.00772   4.22  126.0   \n",
       "...                 ...   ...    ...             ...    ...    ...   \n",
       "7102083         0.00589  1.20   93.5         0.00270   1.58  102.0   \n",
       "7102084         0.00401  2.00   92.1         0.00764   3.00   99.1   \n",
       "7102085         0.00430  1.32   91.8         0.00270   1.33  100.0   \n",
       "7102086         0.00280  2.70   90.7         0.00564   1.45  100.0   \n",
       "7102087         0.00332  3.17   91.7         0.00432   2.10  102.0   \n",
       "\n",
       "         dwelling_time3   sd3  mean3          gene_id  label  \n",
       "0               0.00797  8.75   83.7  ENSG00000004059      0  \n",
       "1               0.00863  6.20   80.0  ENSG00000004059      0  \n",
       "2               0.01590  2.89   78.7  ENSG00000004059      0  \n",
       "3               0.00498  2.63   80.0  ENSG00000004059      0  \n",
       "4               0.00474  5.84   80.9  ENSG00000004059      0  \n",
       "...                 ...   ...    ...              ...    ...  \n",
       "7102083         0.01530  2.19   92.3  ENSG00000284707      0  \n",
       "7102084         0.00469  1.37   92.6  ENSG00000284707      0  \n",
       "7102085         0.00550  1.37   93.9  ENSG00000284707      0  \n",
       "7102086         0.01600  1.90   92.3  ENSG00000284707      0  \n",
       "7102087         0.01320  1.72   93.5  ENSG00000284707      0  \n",
       "\n",
       "[7102088 rows x 14 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_df = df1info.iloc[: , 1:]\n",
    "og_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "805b5019",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_df1 = og_df.groupby(['transcript_id', 'transcript_position']).size().reset_index(name='read_count')\n",
    "count_df2 = og_df.groupby(['transcript_id']).size().reset_index(name='expression_count') #most genes are likely to produce unique transcripts\n",
    "df = og_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "df = df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'sum'\n",
    "}\n",
    "\n",
    "summary_df = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "\n",
    "summary_df = summary_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "summary_df = summary_df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "summary_df['label_percentage'] = summary_df['label'] / summary_df['read_count'] * 100\n",
    "summary_df = summary_df.rename(columns={'label': 'label_sum'})\n",
    "summary_df['label'] = summary_df['label_percentage'].apply(lambda x: 0 if x == 0 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "87c0abb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'first'\n",
    "}\n",
    "\n",
    "summary_df1 = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "summary_df1['label'].equals(summary_df['label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "a71bd5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df = summary_df.drop(columns=['label_percentage', 'label_sum'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "66ac012a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to convert features\n",
    "def sequence_to_features(seq):\n",
    "    \"\"\"\n",
    "    Convert a 7-letter sequence into binary features for the first and last letters.\n",
    "    \n",
    "    Parameters:\n",
    "    - seq: A string of 7 letters (A, T, C, G)\n",
    "    \n",
    "    Returns:\n",
    "    - A dictionary with binary features for the first and last letters.\n",
    "    \"\"\"\n",
    "    letters = ['A', 'T', 'C', 'G']\n",
    "    features = {}\n",
    "    \n",
    "    first_char = seq[0]\n",
    "    last_char = seq[-1]\n",
    "    \n",
    "    for letter in letters:\n",
    "        features[f\"1{letter}\"] = first_char == letter\n",
    "        features[f\"7{letter}\"] = last_char == letter\n",
    "            \n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "b7d24eda",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = summary_df.copy()\n",
    "\n",
    "feature_dicts = new_df['combined nucleotides'].apply(sequence_to_features)\n",
    "features_df = feature_dicts.apply(pd.Series)\n",
    "new_df = pd.concat([new_df, features_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "73a6b413",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5-mer\n",
       "GGACT    897\n",
       "GAACT    474\n",
       "GGACA    399\n",
       "AGACT    330\n",
       "GGACC    310\n",
       "TGACT    267\n",
       "AAACT    147\n",
       "GAACA     98\n",
       "AGACA     87\n",
       "GAACC     82\n",
       "TGACC     58\n",
       "TGACA     42\n",
       "TAACT     41\n",
       "AGACC     28\n",
       "AAACA     23\n",
       "AAACC      6\n",
       "TAACA      1\n",
       "TAACC      1\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df['5-mer'] = new_df['combined nucleotides'].apply(lambda x: x[1:-1])\n",
    "new_df.groupby(['5-mer'])['label'].sum().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "eb863391",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>label</th>\n",
       "      <th>1A</th>\n",
       "      <th>7A</th>\n",
       "      <th>1T</th>\n",
       "      <th>7T</th>\n",
       "      <th>1C</th>\n",
       "      <th>7C</th>\n",
       "      <th>1G</th>\n",
       "      <th>7G</th>\n",
       "      <th>5-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>5</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>5</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>5</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>5</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>5</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>3421</td>\n",
       "      <td>1348</td>\n",
       "      <td>1769</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>3421</td>\n",
       "      <td>1429</td>\n",
       "      <td>1769</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>3421</td>\n",
       "      <td>1531</td>\n",
       "      <td>1769</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>3421</td>\n",
       "      <td>1537</td>\n",
       "      <td>1769</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>3421</td>\n",
       "      <td>1693</td>\n",
       "      <td>1769</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                  0                  244        5              AAGACCA   \n",
       "1                  0                  261        5              CAAACTG   \n",
       "2                  0                  316        5              GAAACAG   \n",
       "3                  0                  332        5              AGAACAT   \n",
       "4                  0                  368        5              AGGACAA   \n",
       "...              ...                  ...      ...                  ...   \n",
       "67315           3421                 1348     1769              GGGACAT   \n",
       "67316           3421                 1429     1769              CTGACAC   \n",
       "67317           3421                 1531     1769              TGGACAC   \n",
       "67318           3421                 1537     1769              CTGACCA   \n",
       "67319           3421                 1693     1769              TTGACAT   \n",
       "\n",
       "       dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0            0.008722  3.990545  123.090909        0.010218  6.716545   \n",
       "1            0.006292  2.849892  109.027711        0.007084  3.197289   \n",
       "2            0.007237  2.782710  105.440237        0.007050  3.463964   \n",
       "3            0.010299  6.042386  129.304348        0.008195  2.733641   \n",
       "4            0.011535  6.031751  118.000000        0.012108  5.748927   \n",
       "...               ...       ...         ...             ...       ...   \n",
       "67315        0.010009  3.119023  118.167630        0.008217  4.816532   \n",
       "67316        0.008481  4.634631  110.052349        0.012667  8.901409   \n",
       "67317        0.007289  3.702195  113.851562        0.007072  4.365023   \n",
       "67318        0.006971  3.148815  109.843704        0.006916  6.331704   \n",
       "67319        0.006879  3.576822  105.493458        0.007417  8.748692   \n",
       "\n",
       "            mean2  ...  label     1A     7A     1T     7T     1C     7C  \\\n",
       "0      125.503030  ...      0   True   True  False  False  False  False   \n",
       "1      107.190964  ...      0  False  False  False  False   True  False   \n",
       "2       98.539053  ...      0  False  False  False  False  False  False   \n",
       "3       97.320652  ...      0   True  False  False   True  False  False   \n",
       "4      121.485876  ...      0   True   True  False  False  False  False   \n",
       "...           ...  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "67315  117.213873  ...      1  False  False  False   True  False  False   \n",
       "67316  114.590604  ...      0  False  False  False  False   True   True   \n",
       "67317  114.125000  ...      1  False  False   True  False  False   True   \n",
       "67318  122.340741  ...      0  False   True  False  False   True  False   \n",
       "67319  113.588785  ...      0  False  False   True   True  False  False   \n",
       "\n",
       "          1G     7G  5-mer  \n",
       "0      False  False  AGACC  \n",
       "1      False   True  AAACT  \n",
       "2       True   True  AAACA  \n",
       "3      False  False  GAACA  \n",
       "4      False  False  GGACA  \n",
       "...      ...    ...    ...  \n",
       "67315   True  False  GGACA  \n",
       "67316  False  False  TGACA  \n",
       "67317  False  False  GGACA  \n",
       "67318  False  False  TGACC  \n",
       "67319  False  False  TGACA  \n",
       "\n",
       "[67320 rows x 25 columns]"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "combined = new_df.copy()\n",
    "\n",
    "label_encoder = LabelEncoder()\n",
    "combined['transcript_id'] = label_encoder.fit_transform(combined['transcript_id'])\n",
    "combined['gene_id'] = label_encoder.fit_transform(combined['gene_id'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "fda7aeb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>5</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>5</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>5</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>5</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>5</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>3421</td>\n",
       "      <td>1348</td>\n",
       "      <td>1769</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>3421</td>\n",
       "      <td>1429</td>\n",
       "      <td>1769</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>3421</td>\n",
       "      <td>1531</td>\n",
       "      <td>1769</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>3421</td>\n",
       "      <td>1537</td>\n",
       "      <td>1769</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>3421</td>\n",
       "      <td>1693</td>\n",
       "      <td>1769</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                  0                  244        5              AAGACCA   \n",
       "1                  0                  261        5              CAAACTG   \n",
       "2                  0                  316        5              GAAACAG   \n",
       "3                  0                  332        5              AGAACAT   \n",
       "4                  0                  368        5              AGGACAA   \n",
       "...              ...                  ...      ...                  ...   \n",
       "67315           3421                 1348     1769              GGGACAT   \n",
       "67316           3421                 1429     1769              CTGACAC   \n",
       "67317           3421                 1531     1769              TGGACAC   \n",
       "67318           3421                 1537     1769              CTGACCA   \n",
       "67319           3421                 1693     1769              TTGACAT   \n",
       "\n",
       "       dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0            0.008722  3.990545  123.090909        0.010218  6.716545   \n",
       "1            0.006292  2.849892  109.027711        0.007084  3.197289   \n",
       "2            0.007237  2.782710  105.440237        0.007050  3.463964   \n",
       "3            0.010299  6.042386  129.304348        0.008195  2.733641   \n",
       "4            0.011535  6.031751  118.000000        0.012108  5.748927   \n",
       "...               ...       ...         ...             ...       ...   \n",
       "67315        0.010009  3.119023  118.167630        0.008217  4.816532   \n",
       "67316        0.008481  4.634631  110.052349        0.012667  8.901409   \n",
       "67317        0.007289  3.702195  113.851562        0.007072  4.365023   \n",
       "67318        0.006971  3.148815  109.843704        0.006916  6.331704   \n",
       "67319        0.006879  3.576822  105.493458        0.007417  8.748692   \n",
       "\n",
       "            mean2  ...  5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  \\\n",
       "0      125.503030  ...        False        False        False        False   \n",
       "1      107.190964  ...        False        False        False        False   \n",
       "2       98.539053  ...        False        False        False        False   \n",
       "3       97.320652  ...        False        False        False        False   \n",
       "4      121.485876  ...        False         True        False        False   \n",
       "...           ...  ...          ...          ...          ...          ...   \n",
       "67315  117.213873  ...        False         True        False        False   \n",
       "67316  114.590604  ...        False        False        False        False   \n",
       "67317  114.125000  ...        False         True        False        False   \n",
       "67318  122.340741  ...        False        False        False        False   \n",
       "67319  113.588785  ...        False        False        False        False   \n",
       "\n",
       "       5-mer_TAACA  5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  \\\n",
       "0            False        False        False        False        False   \n",
       "1            False        False        False        False        False   \n",
       "2            False        False        False        False        False   \n",
       "3            False        False        False        False        False   \n",
       "4            False        False        False        False        False   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "67315        False        False        False        False        False   \n",
       "67316        False        False        False         True        False   \n",
       "67317        False        False        False        False        False   \n",
       "67318        False        False        False        False         True   \n",
       "67319        False        False        False         True        False   \n",
       "\n",
       "       5-mer_TGACT  \n",
       "0            False  \n",
       "1            False  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "...            ...  \n",
       "67315        False  \n",
       "67316        False  \n",
       "67317        False  \n",
       "67318        False  \n",
       "67319        False  \n",
       "\n",
       "[67320 rows x 42 columns]"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = pd.get_dummies(combined, columns=['5-mer'], drop_first=False)\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "9dae6582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>5</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>0.006773</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>5</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>0.007168</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>5</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>0.007141</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>5</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>5</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>0.009484</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>3421</td>\n",
       "      <td>1348</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>3421</td>\n",
       "      <td>1429</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>0.005008</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>3421</td>\n",
       "      <td>1531</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>0.006613</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>3421</td>\n",
       "      <td>1537</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>0.006476</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>3421</td>\n",
       "      <td>1693</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  transcript_position  gene_id  dwelling_time1       sd1  \\\n",
       "0                  0                  244        5        0.008722  3.990545   \n",
       "1                  0                  261        5        0.006292  2.849892   \n",
       "2                  0                  316        5        0.007237  2.782710   \n",
       "3                  0                  332        5        0.010299  6.042386   \n",
       "4                  0                  368        5        0.011535  6.031751   \n",
       "...              ...                  ...      ...             ...       ...   \n",
       "67315           3421                 1348     1769        0.010009  3.119023   \n",
       "67316           3421                 1429     1769        0.008481  4.634631   \n",
       "67317           3421                 1531     1769        0.007289  3.702195   \n",
       "67318           3421                 1537     1769        0.006971  3.148815   \n",
       "67319           3421                 1693     1769        0.006879  3.576822   \n",
       "\n",
       "            mean1  dwelling_time2       sd2       mean2  dwelling_time3  ...  \\\n",
       "0      123.090909        0.010218  6.716545  125.503030        0.006773  ...   \n",
       "1      109.027711        0.007084  3.197289  107.190964        0.007168  ...   \n",
       "2      105.440237        0.007050  3.463964   98.539053        0.007141  ...   \n",
       "3      129.304348        0.008195  2.733641   97.320652        0.005806  ...   \n",
       "4      118.000000        0.012108  5.748927  121.485876        0.009484  ...   \n",
       "...           ...             ...       ...         ...             ...  ...   \n",
       "67315  118.167630        0.008217  4.816532  117.213873        0.007304  ...   \n",
       "67316  110.052349        0.012667  8.901409  114.590604        0.005008  ...   \n",
       "67317  113.851562        0.007072  4.365023  114.125000        0.006613  ...   \n",
       "67318  109.843704        0.006916  6.331704  122.340741        0.006476  ...   \n",
       "67319  105.493458        0.007417  8.748692  113.588785        0.007295  ...   \n",
       "\n",
       "       5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  5-mer_TAACA  \\\n",
       "0            False        False        False        False        False   \n",
       "1            False        False        False        False        False   \n",
       "2            False        False        False        False        False   \n",
       "3            False        False        False        False        False   \n",
       "4            False         True        False        False        False   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "67315        False         True        False        False        False   \n",
       "67316        False        False        False        False        False   \n",
       "67317        False         True        False        False        False   \n",
       "67318        False        False        False        False        False   \n",
       "67319        False        False        False        False        False   \n",
       "\n",
       "       5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  5-mer_TGACT  \n",
       "0            False        False        False        False        False  \n",
       "1            False        False        False        False        False  \n",
       "2            False        False        False        False        False  \n",
       "3            False        False        False        False        False  \n",
       "4            False        False        False        False        False  \n",
       "...            ...          ...          ...          ...          ...  \n",
       "67315        False        False        False        False        False  \n",
       "67316        False        False         True        False        False  \n",
       "67317        False        False        False        False        False  \n",
       "67318        False        False        False         True        False  \n",
       "67319        False        False         True        False        False  \n",
       "\n",
       "[67320 rows x 41 columns]"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = combined.drop(columns=['combined nucleotides'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "bfda3acd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "1311    192\n",
       "307     182\n",
       "250     181\n",
       "1506    177\n",
       "2120    173\n",
       "       ... \n",
       "600       1\n",
       "2327      1\n",
       "748       1\n",
       "2329      1\n",
       "2351      1\n",
       "Name: count, Length: 2501, dtype: int64"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.value_counts('gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "da29573b",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = combined.drop('label', axis=1)\n",
    "y = combined['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "9b317395",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from imblearn.under_sampling import TomekLinks, RandomUnderSampler\n",
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.over_sampling import SMOTE, ADASYN\n",
    "from imblearn.combine import SMOTETomek\n",
    "\n",
    "# ADASYN followed by Tomek Links Pipeline\n",
    "adasyn_tomek = Pipeline([\n",
    "    ('adasyn', ADASYN(random_state=42)),\n",
    "    ('tomek', TomekLinks())\n",
    "])\n",
    "\n",
    "# Dictionary of resampling methods\n",
    "resampling_methods = {\n",
    "    'RandomUnderSampler': RandomUnderSampler(random_state=42),\n",
    "    'TomekLinks': TomekLinks(),\n",
    "    'SMOTE': SMOTE(random_state=42),\n",
    "    'ADASYN': ADASYN(random_state=42),\n",
    "    'SMOTETomek': SMOTETomek(random_state=42),\n",
    "    'ADASYN_Tomek': adasyn_tomek\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "feecf33e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import xgboost as xgb\n",
    "\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, precision_recall_curve\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X1_min_features = X.drop(columns=['gene_id', 'transcript_id', 'transcript_position'])\n",
    "X2_min_features = X.drop(columns=['gene_id', 'transcript_id'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "3514f359",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>read_count</th>\n",
       "      <th>expression_count</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>0.006773</td>\n",
       "      <td>3.999745</td>\n",
       "      <td>80.060606</td>\n",
       "      <td>165</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>0.007168</td>\n",
       "      <td>2.686892</td>\n",
       "      <td>94.572289</td>\n",
       "      <td>166</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>0.007141</td>\n",
       "      <td>2.083775</td>\n",
       "      <td>88.985799</td>\n",
       "      <td>169</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>2.252565</td>\n",
       "      <td>89.350000</td>\n",
       "      <td>184</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>0.009484</td>\n",
       "      <td>3.958192</td>\n",
       "      <td>85.220904</td>\n",
       "      <td>177</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>3.959769</td>\n",
       "      <td>82.101734</td>\n",
       "      <td>173</td>\n",
       "      <td>2880</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>0.005008</td>\n",
       "      <td>3.505034</td>\n",
       "      <td>81.248322</td>\n",
       "      <td>149</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>0.006613</td>\n",
       "      <td>2.339922</td>\n",
       "      <td>84.251563</td>\n",
       "      <td>128</td>\n",
       "      <td>2880</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>0.006476</td>\n",
       "      <td>2.578881</td>\n",
       "      <td>82.337778</td>\n",
       "      <td>135</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>2.689458</td>\n",
       "      <td>79.108411</td>\n",
       "      <td>107</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id  transcript_position          gene_id  \\\n",
       "0      ENST00000000233                  244  ENSG00000004059   \n",
       "1      ENST00000000233                  261  ENSG00000004059   \n",
       "2      ENST00000000233                  316  ENSG00000004059   \n",
       "3      ENST00000000233                  332  ENSG00000004059   \n",
       "4      ENST00000000233                  368  ENSG00000004059   \n",
       "...                ...                  ...              ...   \n",
       "67315  ENST00000641834                 1348  ENSG00000167747   \n",
       "67316  ENST00000641834                 1429  ENSG00000167747   \n",
       "67317  ENST00000641834                 1531  ENSG00000167747   \n",
       "67318  ENST00000641834                 1537  ENSG00000167747   \n",
       "67319  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "      combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                  AAGACCA        0.008722  3.990545  123.090909   \n",
       "1                  CAAACTG        0.006292  2.849892  109.027711   \n",
       "2                  GAAACAG        0.007237  2.782710  105.440237   \n",
       "3                  AGAACAT        0.010299  6.042386  129.304348   \n",
       "4                  AGGACAA        0.011535  6.031751  118.000000   \n",
       "...                    ...             ...       ...         ...   \n",
       "67315              GGGACAT        0.010009  3.119023  118.167630   \n",
       "67316              CTGACAC        0.008481  4.634631  110.052349   \n",
       "67317              TGGACAC        0.007289  3.702195  113.851562   \n",
       "67318              CTGACCA        0.006971  3.148815  109.843704   \n",
       "67319              TTGACAT        0.006879  3.576822  105.493458   \n",
       "\n",
       "       dwelling_time2       sd2       mean2  dwelling_time3       sd3  \\\n",
       "0            0.010218  6.716545  125.503030        0.006773  3.999745   \n",
       "1            0.007084  3.197289  107.190964        0.007168  2.686892   \n",
       "2            0.007050  3.463964   98.539053        0.007141  2.083775   \n",
       "3            0.008195  2.733641   97.320652        0.005806  2.252565   \n",
       "4            0.012108  5.748927  121.485876        0.009484  3.958192   \n",
       "...               ...       ...         ...             ...       ...   \n",
       "67315        0.008217  4.816532  117.213873        0.007304  3.959769   \n",
       "67316        0.012667  8.901409  114.590604        0.005008  3.505034   \n",
       "67317        0.007072  4.365023  114.125000        0.006613  2.339922   \n",
       "67318        0.006916  6.331704  122.340741        0.006476  2.578881   \n",
       "67319        0.007417  8.748692  113.588785        0.007295  2.689458   \n",
       "\n",
       "           mean3  read_count  expression_count  label  \n",
       "0      80.060606         165              3159      0  \n",
       "1      94.572289         166              3159      0  \n",
       "2      88.985799         169              3159      0  \n",
       "3      89.350000         184              3159      0  \n",
       "4      85.220904         177              3159      0  \n",
       "...          ...         ...               ...    ...  \n",
       "67315  82.101734         173              2880      1  \n",
       "67316  81.248322         149              2880      0  \n",
       "67317  84.251563         128              2880      1  \n",
       "67318  82.337778         135              2880      0  \n",
       "67319  79.108411         107              2880      0  \n",
       "\n",
       "[67320 rows x 16 columns]"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "ff2c6f6e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>244</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>0.006773</td>\n",
       "      <td>3.999745</td>\n",
       "      <td>80.060606</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>261</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>0.007168</td>\n",
       "      <td>2.686892</td>\n",
       "      <td>94.572289</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>316</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>0.007141</td>\n",
       "      <td>2.083775</td>\n",
       "      <td>88.985799</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>332</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>2.252565</td>\n",
       "      <td>89.350000</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>368</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>0.009484</td>\n",
       "      <td>3.958192</td>\n",
       "      <td>85.220904</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>1348</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>3.959769</td>\n",
       "      <td>82.101734</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>1429</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>0.005008</td>\n",
       "      <td>3.505034</td>\n",
       "      <td>81.248322</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>1531</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>0.006613</td>\n",
       "      <td>2.339922</td>\n",
       "      <td>84.251563</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>1537</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>0.006476</td>\n",
       "      <td>2.578881</td>\n",
       "      <td>82.337778</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>1693</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>2.689458</td>\n",
       "      <td>79.108411</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_position  dwelling_time1       sd1       mean1  \\\n",
       "0                      244        0.008722  3.990545  123.090909   \n",
       "1                      261        0.006292  2.849892  109.027711   \n",
       "2                      316        0.007237  2.782710  105.440237   \n",
       "3                      332        0.010299  6.042386  129.304348   \n",
       "4                      368        0.011535  6.031751  118.000000   \n",
       "...                    ...             ...       ...         ...   \n",
       "67315                 1348        0.010009  3.119023  118.167630   \n",
       "67316                 1429        0.008481  4.634631  110.052349   \n",
       "67317                 1531        0.007289  3.702195  113.851562   \n",
       "67318                 1537        0.006971  3.148815  109.843704   \n",
       "67319                 1693        0.006879  3.576822  105.493458   \n",
       "\n",
       "       dwelling_time2       sd2       mean2  dwelling_time3       sd3  \\\n",
       "0            0.010218  6.716545  125.503030        0.006773  3.999745   \n",
       "1            0.007084  3.197289  107.190964        0.007168  2.686892   \n",
       "2            0.007050  3.463964   98.539053        0.007141  2.083775   \n",
       "3            0.008195  2.733641   97.320652        0.005806  2.252565   \n",
       "4            0.012108  5.748927  121.485876        0.009484  3.958192   \n",
       "...               ...       ...         ...             ...       ...   \n",
       "67315        0.008217  4.816532  117.213873        0.007304  3.959769   \n",
       "67316        0.012667  8.901409  114.590604        0.005008  3.505034   \n",
       "67317        0.007072  4.365023  114.125000        0.006613  2.339922   \n",
       "67318        0.006916  6.331704  122.340741        0.006476  2.578881   \n",
       "67319        0.007417  8.748692  113.588785        0.007295  2.689458   \n",
       "\n",
       "           mean3  ...  5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  \\\n",
       "0      80.060606  ...        False        False        False        False   \n",
       "1      94.572289  ...        False        False        False        False   \n",
       "2      88.985799  ...        False        False        False        False   \n",
       "3      89.350000  ...        False        False        False        False   \n",
       "4      85.220904  ...        False         True        False        False   \n",
       "...          ...  ...          ...          ...          ...          ...   \n",
       "67315  82.101734  ...        False         True        False        False   \n",
       "67316  81.248322  ...        False        False        False        False   \n",
       "67317  84.251563  ...        False         True        False        False   \n",
       "67318  82.337778  ...        False        False        False        False   \n",
       "67319  79.108411  ...        False        False        False        False   \n",
       "\n",
       "       5-mer_TAACA  5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  \\\n",
       "0            False        False        False        False        False   \n",
       "1            False        False        False        False        False   \n",
       "2            False        False        False        False        False   \n",
       "3            False        False        False        False        False   \n",
       "4            False        False        False        False        False   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "67315        False        False        False        False        False   \n",
       "67316        False        False        False         True        False   \n",
       "67317        False        False        False        False        False   \n",
       "67318        False        False        False        False         True   \n",
       "67319        False        False        False         True        False   \n",
       "\n",
       "       5-mer_TGACT  \n",
       "0            False  \n",
       "1            False  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "...            ...  \n",
       "67315        False  \n",
       "67316        False  \n",
       "67317        False  \n",
       "67318        False  \n",
       "67319        False  \n",
       "\n",
       "[67320 rows x 38 columns]"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X2_min_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "3f289d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "model1 = pickle.load(open('xgb_model_1.pkl', 'rb'))\n",
    "model2 = pickle.load(open('xgb_model_2.pkl','rb'))\n",
    "X1_predicted = model1.predict_proba(X1_min_features)[:, 1]\n",
    "X2_predicted = model2.predict_proba(X2_min_features)[:, 1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "a8d9de05",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final1 = summary_df\n",
    "summary_df_final1['score'] =X1_predicted.tolist()\n",
    "summary_df_final2 = summary_df\n",
    "summary_df_final2['score']= X2_predicted.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "6dd02bbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>read_count</th>\n",
       "      <th>expression_count</th>\n",
       "      <th>label</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>3.990545</td>\n",
       "      <td>123.090909</td>\n",
       "      <td>0.010218</td>\n",
       "      <td>6.716545</td>\n",
       "      <td>125.503030</td>\n",
       "      <td>0.006773</td>\n",
       "      <td>3.999745</td>\n",
       "      <td>80.060606</td>\n",
       "      <td>165</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.074175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006292</td>\n",
       "      <td>2.849892</td>\n",
       "      <td>109.027711</td>\n",
       "      <td>0.007084</td>\n",
       "      <td>3.197289</td>\n",
       "      <td>107.190964</td>\n",
       "      <td>0.007168</td>\n",
       "      <td>2.686892</td>\n",
       "      <td>94.572289</td>\n",
       "      <td>166</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.084994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007237</td>\n",
       "      <td>2.782710</td>\n",
       "      <td>105.440237</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>3.463964</td>\n",
       "      <td>98.539053</td>\n",
       "      <td>0.007141</td>\n",
       "      <td>2.083775</td>\n",
       "      <td>88.985799</td>\n",
       "      <td>169</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.046752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010299</td>\n",
       "      <td>6.042386</td>\n",
       "      <td>129.304348</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>2.733641</td>\n",
       "      <td>97.320652</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>2.252565</td>\n",
       "      <td>89.350000</td>\n",
       "      <td>184</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.049987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.011535</td>\n",
       "      <td>6.031751</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>5.748927</td>\n",
       "      <td>121.485876</td>\n",
       "      <td>0.009484</td>\n",
       "      <td>3.958192</td>\n",
       "      <td>85.220904</td>\n",
       "      <td>177</td>\n",
       "      <td>3159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.117183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67315</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>3.119023</td>\n",
       "      <td>118.167630</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>4.816532</td>\n",
       "      <td>117.213873</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>3.959769</td>\n",
       "      <td>82.101734</td>\n",
       "      <td>173</td>\n",
       "      <td>2880</td>\n",
       "      <td>1</td>\n",
       "      <td>0.044916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>4.634631</td>\n",
       "      <td>110.052349</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>8.901409</td>\n",
       "      <td>114.590604</td>\n",
       "      <td>0.005008</td>\n",
       "      <td>3.505034</td>\n",
       "      <td>81.248322</td>\n",
       "      <td>149</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "      <td>0.060924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67317</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.007289</td>\n",
       "      <td>3.702195</td>\n",
       "      <td>113.851562</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>4.365023</td>\n",
       "      <td>114.125000</td>\n",
       "      <td>0.006613</td>\n",
       "      <td>2.339922</td>\n",
       "      <td>84.251563</td>\n",
       "      <td>128</td>\n",
       "      <td>2880</td>\n",
       "      <td>1</td>\n",
       "      <td>0.047084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.006971</td>\n",
       "      <td>3.148815</td>\n",
       "      <td>109.843704</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>6.331704</td>\n",
       "      <td>122.340741</td>\n",
       "      <td>0.006476</td>\n",
       "      <td>2.578881</td>\n",
       "      <td>82.337778</td>\n",
       "      <td>135</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "      <td>0.113907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67319</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>3.576822</td>\n",
       "      <td>105.493458</td>\n",
       "      <td>0.007417</td>\n",
       "      <td>8.748692</td>\n",
       "      <td>113.588785</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>2.689458</td>\n",
       "      <td>79.108411</td>\n",
       "      <td>107</td>\n",
       "      <td>2880</td>\n",
       "      <td>0</td>\n",
       "      <td>0.047013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67320 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id  transcript_position          gene_id  \\\n",
       "0      ENST00000000233                  244  ENSG00000004059   \n",
       "1      ENST00000000233                  261  ENSG00000004059   \n",
       "2      ENST00000000233                  316  ENSG00000004059   \n",
       "3      ENST00000000233                  332  ENSG00000004059   \n",
       "4      ENST00000000233                  368  ENSG00000004059   \n",
       "...                ...                  ...              ...   \n",
       "67315  ENST00000641834                 1348  ENSG00000167747   \n",
       "67316  ENST00000641834                 1429  ENSG00000167747   \n",
       "67317  ENST00000641834                 1531  ENSG00000167747   \n",
       "67318  ENST00000641834                 1537  ENSG00000167747   \n",
       "67319  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "      combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                  AAGACCA        0.008722  3.990545  123.090909   \n",
       "1                  CAAACTG        0.006292  2.849892  109.027711   \n",
       "2                  GAAACAG        0.007237  2.782710  105.440237   \n",
       "3                  AGAACAT        0.010299  6.042386  129.304348   \n",
       "4                  AGGACAA        0.011535  6.031751  118.000000   \n",
       "...                    ...             ...       ...         ...   \n",
       "67315              GGGACAT        0.010009  3.119023  118.167630   \n",
       "67316              CTGACAC        0.008481  4.634631  110.052349   \n",
       "67317              TGGACAC        0.007289  3.702195  113.851562   \n",
       "67318              CTGACCA        0.006971  3.148815  109.843704   \n",
       "67319              TTGACAT        0.006879  3.576822  105.493458   \n",
       "\n",
       "       dwelling_time2       sd2       mean2  dwelling_time3       sd3  \\\n",
       "0            0.010218  6.716545  125.503030        0.006773  3.999745   \n",
       "1            0.007084  3.197289  107.190964        0.007168  2.686892   \n",
       "2            0.007050  3.463964   98.539053        0.007141  2.083775   \n",
       "3            0.008195  2.733641   97.320652        0.005806  2.252565   \n",
       "4            0.012108  5.748927  121.485876        0.009484  3.958192   \n",
       "...               ...       ...         ...             ...       ...   \n",
       "67315        0.008217  4.816532  117.213873        0.007304  3.959769   \n",
       "67316        0.012667  8.901409  114.590604        0.005008  3.505034   \n",
       "67317        0.007072  4.365023  114.125000        0.006613  2.339922   \n",
       "67318        0.006916  6.331704  122.340741        0.006476  2.578881   \n",
       "67319        0.007417  8.748692  113.588785        0.007295  2.689458   \n",
       "\n",
       "           mean3  read_count  expression_count  label     score  \n",
       "0      80.060606         165              3159      0  0.074175  \n",
       "1      94.572289         166              3159      0  0.084994  \n",
       "2      88.985799         169              3159      0  0.046752  \n",
       "3      89.350000         184              3159      0  0.049987  \n",
       "4      85.220904         177              3159      0  0.117183  \n",
       "...          ...         ...               ...    ...       ...  \n",
       "67315  82.101734         173              2880      1  0.044916  \n",
       "67316  81.248322         149              2880      0  0.060924  \n",
       "67317  84.251563         128              2880      1  0.047084  \n",
       "67318  82.337778         135              2880      0  0.113907  \n",
       "67319  79.108411         107              2880      0  0.047013  \n",
       "\n",
       "[67320 rows x 17 columns]"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df_final1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "b9371673",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final1[[\"transcript_id\",\"transcript_position\",\"score\"]].to_csv('dataset1csv1.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "f01b72c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final2[[\"transcript_id\",\"transcript_position\",\"score\"]].to_csv('dataset1csv2.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "d904e9e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0info = pd.merge(df0, data, on=['transcript_id', 'transcript_position'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "29f4d829",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_df = df0info.iloc[: , 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "d8e9b7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_df1 = og_df.groupby(['transcript_id', 'transcript_position']).size().reset_index(name='read_count')\n",
    "count_df2 = og_df.groupby(['transcript_id']).size().reset_index(name='expression_count') #most genes are likely to produce unique transcripts\n",
    "df = og_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "df = df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'sum'\n",
    "}\n",
    "\n",
    "summary_df = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "\n",
    "summary_df = summary_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "summary_df = summary_df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "summary_df['label_percentage'] = summary_df['label'] / summary_df['read_count'] * 100\n",
    "summary_df = summary_df.rename(columns={'label': 'label_sum'})\n",
    "summary_df['label'] = summary_df['label_percentage'].apply(lambda x: 0 if x == 0 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "13652750",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'first'\n",
    "}\n",
    "\n",
    "summary_df1 = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "summary_df1['label'].equals(summary_df['label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "a7272bb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>read_count</th>\n",
       "      <th>expression_count</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>185</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>172</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>185</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>200</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>198</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>4.005616</td>\n",
       "      <td>82.004110</td>\n",
       "      <td>73</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>3.644638</td>\n",
       "      <td>80.497101</td>\n",
       "      <td>69</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>2.181563</td>\n",
       "      <td>84.190625</td>\n",
       "      <td>64</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>2.540877</td>\n",
       "      <td>82.289474</td>\n",
       "      <td>57</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>2.576731</td>\n",
       "      <td>78.536538</td>\n",
       "      <td>52</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_position          gene_id  \\\n",
       "0       ENST00000000233                  244  ENSG00000004059   \n",
       "1       ENST00000000233                  261  ENSG00000004059   \n",
       "2       ENST00000000233                  316  ENSG00000004059   \n",
       "3       ENST00000000233                  332  ENSG00000004059   \n",
       "4       ENST00000000233                  368  ENSG00000004059   \n",
       "...                 ...                  ...              ...   \n",
       "121833  ENST00000641834                 1348  ENSG00000167747   \n",
       "121834  ENST00000641834                 1429  ENSG00000167747   \n",
       "121835  ENST00000641834                 1531  ENSG00000167747   \n",
       "121836  ENST00000641834                 1537  ENSG00000167747   \n",
       "121837  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "       combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                   AAGACCA        0.008264  4.223784  123.702703   \n",
       "1                   CAAACTG        0.006609  3.216424  109.681395   \n",
       "2                   GAAACAG        0.007570  2.940541  105.475676   \n",
       "3                   AGAACAT        0.010620  6.476350  129.355000   \n",
       "4                   AGGACAA        0.010701  6.415051  117.924242   \n",
       "...                     ...             ...       ...         ...   \n",
       "121833              GGGACAT        0.009594  3.294164  118.232877   \n",
       "121834              CTGACAC        0.008393  4.511014  110.969565   \n",
       "121835              TGGACAC        0.008161  3.918438  113.968750   \n",
       "121836              CTGACCA        0.008044  3.191228  109.354386   \n",
       "121837              TTGACAT        0.008788  4.090577  105.807692   \n",
       "\n",
       "        dwelling_time2       sd2       mean2  dwelling_time3       sd3  \\\n",
       "0             0.009373  7.382162  125.913514        0.007345  4.386989   \n",
       "1             0.006813  3.226535  107.889535        0.007710  3.016599   \n",
       "2             0.007416  3.642703   98.947027        0.007555  2.087146   \n",
       "3             0.008632  2.899200   97.836500        0.006101  2.236520   \n",
       "4             0.011479  5.870303  121.954545        0.010019  4.260253   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.007300  4.929726  116.342466        0.006555  4.005616   \n",
       "121834        0.010305  9.105797  114.927536        0.005568  3.644638   \n",
       "121835        0.006877  4.759688  113.562500        0.006410  2.181563   \n",
       "121836        0.007419  6.552982  123.263158        0.006472  2.540877   \n",
       "121837        0.006908  8.702885  113.134615        0.008337  2.576731   \n",
       "\n",
       "            mean3  read_count  expression_count  label  \n",
       "0       80.570270         185              3449      0  \n",
       "1       94.290698         172              3449      0  \n",
       "2       89.364324         185              3449      0  \n",
       "3       89.154000         200              3449      0  \n",
       "4       85.178788         198              3449      0  \n",
       "...           ...         ...               ...    ...  \n",
       "121833  82.004110          73              1323      1  \n",
       "121834  80.497101          69              1323      0  \n",
       "121835  84.190625          64              1323      1  \n",
       "121836  82.289474          57              1323      0  \n",
       "121837  78.536538          52              1323      0  \n",
       "\n",
       "[121838 rows x 16 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = summary_df.drop(columns=['label_percentage', 'label_sum'])\n",
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "68301991",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to convert features\n",
    "def sequence_to_features(seq):\n",
    "    \"\"\"\n",
    "    Convert a 7-letter sequence into binary features for the first and last letters.\n",
    "    \n",
    "    Parameters:\n",
    "    - seq: A string of 7 letters (A, T, C, G)\n",
    "    \n",
    "    Returns:\n",
    "    - A dictionary with binary features for the first and last letters.\n",
    "    \"\"\"\n",
    "    letters = ['A', 'T', 'C', 'G']\n",
    "    features = {}\n",
    "    \n",
    "    first_char = seq[0]\n",
    "    last_char = seq[-1]\n",
    "    \n",
    "    for letter in letters:\n",
    "        features[f\"1{letter}\"] = first_char == letter\n",
    "        features[f\"7{letter}\"] = last_char == letter\n",
    "            \n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "055b423e",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = summary_df.copy()\n",
    "\n",
    "feature_dicts = new_df['combined nucleotides'].apply(sequence_to_features)\n",
    "features_df = feature_dicts.apply(pd.Series)\n",
    "new_df = pd.concat([new_df, features_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "b45b6f5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5-mer\n",
       "GGACT    1440\n",
       "GAACT     783\n",
       "GGACA     691\n",
       "GGACC     552\n",
       "AGACT     525\n",
       "TGACT     430\n",
       "AAACT     255\n",
       "GAACA     187\n",
       "AGACA     147\n",
       "GAACC     124\n",
       "TGACC      90\n",
       "TAACT      71\n",
       "TGACA      68\n",
       "AGACC      60\n",
       "AAACA      38\n",
       "AAACC      11\n",
       "TAACC       2\n",
       "TAACA       1\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df['5-mer'] = new_df['combined nucleotides'].apply(lambda x: x[1:-1])\n",
    "new_df.groupby(['5-mer'])['label'].sum().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "fbbda745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>label</th>\n",
       "      <th>1A</th>\n",
       "      <th>7A</th>\n",
       "      <th>1T</th>\n",
       "      <th>7T</th>\n",
       "      <th>1C</th>\n",
       "      <th>7C</th>\n",
       "      <th>1G</th>\n",
       "      <th>7G</th>\n",
       "      <th>5-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                   0                  244       10              AAGACCA   \n",
       "1                   0                  261       10              CAAACTG   \n",
       "2                   0                  316       10              GAAACAG   \n",
       "3                   0                  332       10              AGAACAT   \n",
       "4                   0                  368       10              AGGACAA   \n",
       "...               ...                  ...      ...                  ...   \n",
       "121833           5332                 1348     2764              GGGACAT   \n",
       "121834           5332                 1429     2764              CTGACAC   \n",
       "121835           5332                 1531     2764              TGGACAC   \n",
       "121836           5332                 1537     2764              CTGACCA   \n",
       "121837           5332                 1693     2764              TTGACAT   \n",
       "\n",
       "        dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0             0.008264  4.223784  123.702703        0.009373  7.382162   \n",
       "1             0.006609  3.216424  109.681395        0.006813  3.226535   \n",
       "2             0.007570  2.940541  105.475676        0.007416  3.642703   \n",
       "3             0.010620  6.476350  129.355000        0.008632  2.899200   \n",
       "4             0.010701  6.415051  117.924242        0.011479  5.870303   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.009594  3.294164  118.232877        0.007300  4.929726   \n",
       "121834        0.008393  4.511014  110.969565        0.010305  9.105797   \n",
       "121835        0.008161  3.918438  113.968750        0.006877  4.759688   \n",
       "121836        0.008044  3.191228  109.354386        0.007419  6.552982   \n",
       "121837        0.008788  4.090577  105.807692        0.006908  8.702885   \n",
       "\n",
       "             mean2  ...  label     1A     7A     1T     7T     1C     7C  \\\n",
       "0       125.913514  ...      0   True   True  False  False  False  False   \n",
       "1       107.889535  ...      0  False  False  False  False   True  False   \n",
       "2        98.947027  ...      0  False  False  False  False  False  False   \n",
       "3        97.836500  ...      0   True  False  False   True  False  False   \n",
       "4       121.954545  ...      0   True   True  False  False  False  False   \n",
       "...            ...  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "121833  116.342466  ...      1  False  False  False   True  False  False   \n",
       "121834  114.927536  ...      0  False  False  False  False   True   True   \n",
       "121835  113.562500  ...      1  False  False   True  False  False   True   \n",
       "121836  123.263158  ...      0  False   True  False  False   True  False   \n",
       "121837  113.134615  ...      0  False  False   True   True  False  False   \n",
       "\n",
       "           1G     7G  5-mer  \n",
       "0       False  False  AGACC  \n",
       "1       False   True  AAACT  \n",
       "2        True   True  AAACA  \n",
       "3       False  False  GAACA  \n",
       "4       False  False  GGACA  \n",
       "...       ...    ...    ...  \n",
       "121833   True  False  GGACA  \n",
       "121834  False  False  TGACA  \n",
       "121835  False  False  GGACA  \n",
       "121836  False  False  TGACC  \n",
       "121837  False  False  TGACA  \n",
       "\n",
       "[121838 rows x 25 columns]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "combined = new_df.copy()\n",
    "\n",
    "label_encoder = LabelEncoder()\n",
    "combined['transcript_id'] = label_encoder.fit_transform(combined['transcript_id'])\n",
    "combined['gene_id'] = label_encoder.fit_transform(combined['gene_id'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "9fb08bfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                   0                  244       10              AAGACCA   \n",
       "1                   0                  261       10              CAAACTG   \n",
       "2                   0                  316       10              GAAACAG   \n",
       "3                   0                  332       10              AGAACAT   \n",
       "4                   0                  368       10              AGGACAA   \n",
       "...               ...                  ...      ...                  ...   \n",
       "121833           5332                 1348     2764              GGGACAT   \n",
       "121834           5332                 1429     2764              CTGACAC   \n",
       "121835           5332                 1531     2764              TGGACAC   \n",
       "121836           5332                 1537     2764              CTGACCA   \n",
       "121837           5332                 1693     2764              TTGACAT   \n",
       "\n",
       "        dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0             0.008264  4.223784  123.702703        0.009373  7.382162   \n",
       "1             0.006609  3.216424  109.681395        0.006813  3.226535   \n",
       "2             0.007570  2.940541  105.475676        0.007416  3.642703   \n",
       "3             0.010620  6.476350  129.355000        0.008632  2.899200   \n",
       "4             0.010701  6.415051  117.924242        0.011479  5.870303   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.009594  3.294164  118.232877        0.007300  4.929726   \n",
       "121834        0.008393  4.511014  110.969565        0.010305  9.105797   \n",
       "121835        0.008161  3.918438  113.968750        0.006877  4.759688   \n",
       "121836        0.008044  3.191228  109.354386        0.007419  6.552982   \n",
       "121837        0.008788  4.090577  105.807692        0.006908  8.702885   \n",
       "\n",
       "             mean2  ...  5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  \\\n",
       "0       125.913514  ...        False        False        False        False   \n",
       "1       107.889535  ...        False        False        False        False   \n",
       "2        98.947027  ...        False        False        False        False   \n",
       "3        97.836500  ...        False        False        False        False   \n",
       "4       121.954545  ...        False         True        False        False   \n",
       "...            ...  ...          ...          ...          ...          ...   \n",
       "121833  116.342466  ...        False         True        False        False   \n",
       "121834  114.927536  ...        False        False        False        False   \n",
       "121835  113.562500  ...        False         True        False        False   \n",
       "121836  123.263158  ...        False        False        False        False   \n",
       "121837  113.134615  ...        False        False        False        False   \n",
       "\n",
       "        5-mer_TAACA  5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  \\\n",
       "0             False        False        False        False        False   \n",
       "1             False        False        False        False        False   \n",
       "2             False        False        False        False        False   \n",
       "3             False        False        False        False        False   \n",
       "4             False        False        False        False        False   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "121833        False        False        False        False        False   \n",
       "121834        False        False        False         True        False   \n",
       "121835        False        False        False        False        False   \n",
       "121836        False        False        False        False         True   \n",
       "121837        False        False        False         True        False   \n",
       "\n",
       "        5-mer_TGACT  \n",
       "0             False  \n",
       "1             False  \n",
       "2             False  \n",
       "3             False  \n",
       "4             False  \n",
       "...             ...  \n",
       "121833        False  \n",
       "121834        False  \n",
       "121835        False  \n",
       "121836        False  \n",
       "121837        False  \n",
       "\n",
       "[121838 rows x 42 columns]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = pd.get_dummies(combined, columns=['5-mer'], drop_first=False)\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "be6127ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id  dwelling_time1       sd1  \\\n",
       "0                   0                  244       10        0.008264  4.223784   \n",
       "1                   0                  261       10        0.006609  3.216424   \n",
       "2                   0                  316       10        0.007570  2.940541   \n",
       "3                   0                  332       10        0.010620  6.476350   \n",
       "4                   0                  368       10        0.010701  6.415051   \n",
       "...               ...                  ...      ...             ...       ...   \n",
       "121833           5332                 1348     2764        0.009594  3.294164   \n",
       "121834           5332                 1429     2764        0.008393  4.511014   \n",
       "121835           5332                 1531     2764        0.008161  3.918438   \n",
       "121836           5332                 1537     2764        0.008044  3.191228   \n",
       "121837           5332                 1693     2764        0.008788  4.090577   \n",
       "\n",
       "             mean1  dwelling_time2       sd2       mean2  dwelling_time3  ...  \\\n",
       "0       123.702703        0.009373  7.382162  125.913514        0.007345  ...   \n",
       "1       109.681395        0.006813  3.226535  107.889535        0.007710  ...   \n",
       "2       105.475676        0.007416  3.642703   98.947027        0.007555  ...   \n",
       "3       129.355000        0.008632  2.899200   97.836500        0.006101  ...   \n",
       "4       117.924242        0.011479  5.870303  121.954545        0.010019  ...   \n",
       "...            ...             ...       ...         ...             ...  ...   \n",
       "121833  118.232877        0.007300  4.929726  116.342466        0.006555  ...   \n",
       "121834  110.969565        0.010305  9.105797  114.927536        0.005568  ...   \n",
       "121835  113.968750        0.006877  4.759688  113.562500        0.006410  ...   \n",
       "121836  109.354386        0.007419  6.552982  123.263158        0.006472  ...   \n",
       "121837  105.807692        0.006908  8.702885  113.134615        0.008337  ...   \n",
       "\n",
       "        5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  5-mer_TAACA  \\\n",
       "0             False        False        False        False        False   \n",
       "1             False        False        False        False        False   \n",
       "2             False        False        False        False        False   \n",
       "3             False        False        False        False        False   \n",
       "4             False         True        False        False        False   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "121833        False         True        False        False        False   \n",
       "121834        False        False        False        False        False   \n",
       "121835        False         True        False        False        False   \n",
       "121836        False        False        False        False        False   \n",
       "121837        False        False        False        False        False   \n",
       "\n",
       "        5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  5-mer_TGACT  \n",
       "0             False        False        False        False        False  \n",
       "1             False        False        False        False        False  \n",
       "2             False        False        False        False        False  \n",
       "3             False        False        False        False        False  \n",
       "4             False        False        False        False        False  \n",
       "...             ...          ...          ...          ...          ...  \n",
       "121833        False        False        False        False        False  \n",
       "121834        False        False         True        False        False  \n",
       "121835        False        False        False        False        False  \n",
       "121836        False        False        False         True        False  \n",
       "121837        False        False         True        False        False  \n",
       "\n",
       "[121838 rows x 41 columns]"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = combined.drop(columns=['combined nucleotides'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "32edd4fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "1528    274\n",
       "977     253\n",
       "2632    250\n",
       "2005    230\n",
       "1929    229\n",
       "       ... \n",
       "857       1\n",
       "863       1\n",
       "887       1\n",
       "2800      1\n",
       "1412      1\n",
       "Name: count, Length: 3852, dtype: int64"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.value_counts('gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "1075f8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = combined.drop('label', axis=1)\n",
    "y = combined['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "e7861982",
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.under_sampling import TomekLinks, RandomUnderSampler\n",
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.over_sampling import SMOTE, ADASYN\n",
    "from imblearn.combine import SMOTETomek\n",
    "\n",
    "# ADASYN followed by Tomek Links Pipeline\n",
    "adasyn_tomek = Pipeline([\n",
    "    ('adasyn', ADASYN(random_state=42)),\n",
    "    ('tomek', TomekLinks())\n",
    "])\n",
    "\n",
    "# Dictionary of resampling methods\n",
    "resampling_methods = {\n",
    "    'RandomUnderSampler': RandomUnderSampler(random_state=42),\n",
    "    'TomekLinks': TomekLinks(),\n",
    "    'SMOTE': SMOTE(random_state=42),\n",
    "    'ADASYN': ADASYN(random_state=42),\n",
    "    'SMOTETomek': SMOTETomek(random_state=42),\n",
    "    'ADASYN_Tomek': adasyn_tomek\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "56fbad39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop 'gene_id' from X\n",
    "X1_min_features = X.drop(columns=['gene_id', 'transcript_id', 'transcript_position'])\n",
    "X2_min_features = X.drop(columns=['gene_id', 'transcript_id'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "cd0367bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "model1 = pickle.load(open('xgb_model_1.pkl', 'rb'))\n",
    "model2 = pickle.load(open('xgb_model_2.pkl','rb'))\n",
    "X1_predicted = model1.predict_proba(X1_min_features)[:, 1]\n",
    "X2_predicted = model2.predict_proba(X2_min_features)[:, 1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "c26b51d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final1 = summary_df\n",
    "summary_df_final1['score'] =X1_predicted.tolist()\n",
    "summary_df_final2 = summary_df\n",
    "summary_df_final2['score']= X2_predicted.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "2b8bb956",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final1[[\"transcript_id\",\"transcript_position\",\"score\"]].to_csv('dataset0csv1.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "88d87b17",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df_final2[[\"transcript_id\",\"transcript_position\",\"score\"]].to_csv('dataset0csv2.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "47ab8332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>2.16</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00640</td>\n",
       "      <td>3.90</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>8.75</td>\n",
       "      <td>83.7</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>4.43</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.01860</td>\n",
       "      <td>10.00</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>6.20</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>3.10</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>8.26</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01590</td>\n",
       "      <td>2.89</td>\n",
       "      <td>78.7</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>4.52</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.01750</td>\n",
       "      <td>8.51</td>\n",
       "      <td>128.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.63</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>2.81</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00772</td>\n",
       "      <td>4.22</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.00474</td>\n",
       "      <td>5.84</td>\n",
       "      <td>80.9</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102083</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>1.20</td>\n",
       "      <td>93.5</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.58</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01530</td>\n",
       "      <td>2.19</td>\n",
       "      <td>92.3</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102084</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>2.00</td>\n",
       "      <td>92.1</td>\n",
       "      <td>0.00764</td>\n",
       "      <td>3.00</td>\n",
       "      <td>99.1</td>\n",
       "      <td>0.00469</td>\n",
       "      <td>1.37</td>\n",
       "      <td>92.6</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102085</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00430</td>\n",
       "      <td>1.32</td>\n",
       "      <td>91.8</td>\n",
       "      <td>0.00270</td>\n",
       "      <td>1.33</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.00550</td>\n",
       "      <td>1.37</td>\n",
       "      <td>93.9</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102086</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00280</td>\n",
       "      <td>2.70</td>\n",
       "      <td>90.7</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>1.45</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.01600</td>\n",
       "      <td>1.90</td>\n",
       "      <td>92.3</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7102087</th>\n",
       "      <td>ENST00000641784</td>\n",
       "      <td>3266</td>\n",
       "      <td>CTAACTT</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>3.17</td>\n",
       "      <td>91.7</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>2.10</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.01320</td>\n",
       "      <td>1.72</td>\n",
       "      <td>93.5</td>\n",
       "      <td>ENSG00000284707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7102088 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id  transcript_position combined nucleotides  \\\n",
       "0        ENST00000000233                  244              AAGACCA   \n",
       "1        ENST00000000233                  244              AAGACCA   \n",
       "2        ENST00000000233                  244              AAGACCA   \n",
       "3        ENST00000000233                  244              AAGACCA   \n",
       "4        ENST00000000233                  244              AAGACCA   \n",
       "...                  ...                  ...                  ...   \n",
       "7102083  ENST00000641784                 3266              CTAACTT   \n",
       "7102084  ENST00000641784                 3266              CTAACTT   \n",
       "7102085  ENST00000641784                 3266              CTAACTT   \n",
       "7102086  ENST00000641784                 3266              CTAACTT   \n",
       "7102087  ENST00000641784                 3266              CTAACTT   \n",
       "\n",
       "         dwelling_time1   sd1  mean1  dwelling_time2    sd2  mean2  \\\n",
       "0               0.00465  2.16  127.0         0.00640   3.90  127.0   \n",
       "1               0.02690  4.43  106.0         0.01860  10.00  123.0   \n",
       "2               0.00432  3.10  108.0         0.01200   8.26  125.0   \n",
       "3               0.00996  4.52  123.0         0.01750   8.51  128.0   \n",
       "4               0.00764  2.81  124.0         0.00772   4.22  126.0   \n",
       "...                 ...   ...    ...             ...    ...    ...   \n",
       "7102083         0.00589  1.20   93.5         0.00270   1.58  102.0   \n",
       "7102084         0.00401  2.00   92.1         0.00764   3.00   99.1   \n",
       "7102085         0.00430  1.32   91.8         0.00270   1.33  100.0   \n",
       "7102086         0.00280  2.70   90.7         0.00564   1.45  100.0   \n",
       "7102087         0.00332  3.17   91.7         0.00432   2.10  102.0   \n",
       "\n",
       "         dwelling_time3   sd3  mean3          gene_id  label  \n",
       "0               0.00797  8.75   83.7  ENSG00000004059      0  \n",
       "1               0.00863  6.20   80.0  ENSG00000004059      0  \n",
       "2               0.01590  2.89   78.7  ENSG00000004059      0  \n",
       "3               0.00498  2.63   80.0  ENSG00000004059      0  \n",
       "4               0.00474  5.84   80.9  ENSG00000004059      0  \n",
       "...                 ...   ...    ...              ...    ...  \n",
       "7102083         0.01530  2.19   92.3  ENSG00000284707      0  \n",
       "7102084         0.00469  1.37   92.6  ENSG00000284707      0  \n",
       "7102085         0.00550  1.37   93.9  ENSG00000284707      0  \n",
       "7102086         0.01600  1.90   92.3  ENSG00000284707      0  \n",
       "7102087         0.01320  1.72   93.5  ENSG00000284707      0  \n",
       "\n",
       "[7102088 rows x 14 columns]"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1info.iloc[: , 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "40f3ce4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.01220</td>\n",
       "      <td>3.99</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00337</td>\n",
       "      <td>4.56</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>4.200</td>\n",
       "      <td>84.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.03020</td>\n",
       "      <td>2.32</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.00443</td>\n",
       "      <td>2.36</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>2.130</td>\n",
       "      <td>79.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>5.55</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>7.04</td>\n",
       "      <td>99.3</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>2.210</td>\n",
       "      <td>86.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>2.10</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>3.90</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>2.180</td>\n",
       "      <td>82.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tx_id_0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAACCT</td>\n",
       "      <td>0.02110</td>\n",
       "      <td>3.49</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>3.80</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.00997</td>\n",
       "      <td>2.180</td>\n",
       "      <td>81.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171935</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00675</td>\n",
       "      <td>8.33</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00759</td>\n",
       "      <td>6.15</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>2.420</td>\n",
       "      <td>91.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171936</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00529</td>\n",
       "      <td>4.81</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>7.42</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.00996</td>\n",
       "      <td>1.940</td>\n",
       "      <td>90.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171937</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.01460</td>\n",
       "      <td>7.72</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>1.21</td>\n",
       "      <td>92.4</td>\n",
       "      <td>0.02150</td>\n",
       "      <td>2.110</td>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171938</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00610</td>\n",
       "      <td>4.21</td>\n",
       "      <td>119.0</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>5.44</td>\n",
       "      <td>94.8</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>2.000</td>\n",
       "      <td>82.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171939</th>\n",
       "      <td>tx_id_6</td>\n",
       "      <td>1880</td>\n",
       "      <td>CGAACAA</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>3.02</td>\n",
       "      <td>96.8</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>1.20</td>\n",
       "      <td>91.5</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>0.913</td>\n",
       "      <td>89.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1171940 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position combined nucleotides  \\\n",
       "0             tx_id_0                    0              AAAACCT   \n",
       "1             tx_id_0                    0              AAAACCT   \n",
       "2             tx_id_0                    0              AAAACCT   \n",
       "3             tx_id_0                    0              AAAACCT   \n",
       "4             tx_id_0                    0              AAAACCT   \n",
       "...               ...                  ...                  ...   \n",
       "1171935       tx_id_6                 1880              CGAACAA   \n",
       "1171936       tx_id_6                 1880              CGAACAA   \n",
       "1171937       tx_id_6                 1880              CGAACAA   \n",
       "1171938       tx_id_6                 1880              CGAACAA   \n",
       "1171939       tx_id_6                 1880              CGAACAA   \n",
       "\n",
       "         dwelling_time1   sd1  mean1  dwelling_time2   sd2  mean2  \\\n",
       "0               0.01220  3.99  106.0         0.00337  4.56  102.0   \n",
       "1               0.03020  2.32  107.0         0.00443  2.36  102.0   \n",
       "2               0.00232  5.55  110.0         0.00664  7.04   99.3   \n",
       "3               0.00465  2.10  104.0         0.00996  3.90  108.0   \n",
       "4               0.02110  3.49  103.0         0.00531  3.80  101.0   \n",
       "...                 ...   ...    ...             ...   ...    ...   \n",
       "1171935         0.00675  8.33  111.0         0.00759  6.15  104.0   \n",
       "1171936         0.00529  4.81  106.0         0.00332  7.42  101.0   \n",
       "1171937         0.01460  7.72  104.0         0.00266  1.21   92.4   \n",
       "1171938         0.00610  4.21  119.0         0.00232  5.44   94.8   \n",
       "1171939         0.00232  3.02   96.8         0.00465  1.20   91.5   \n",
       "\n",
       "         dwelling_time3    sd3  mean3  \n",
       "0               0.00664  4.200   84.2  \n",
       "1               0.00332  2.130   79.2  \n",
       "2               0.00232  2.210   86.6  \n",
       "3               0.00401  2.180   82.2  \n",
       "4               0.00997  2.180   81.2  \n",
       "...                 ...    ...    ...  \n",
       "1171935         0.00896  2.420   91.6  \n",
       "1171936         0.00996  1.940   90.9  \n",
       "1171937         0.02150  2.110   90.0  \n",
       "1171938         0.00199  2.000   82.6  \n",
       "1171939         0.00398  0.913   89.8  \n",
       "\n",
       "[1171940 rows x 12 columns]"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_df = df2.iloc[: , 1:]\n",
    "og_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919cd05d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7670215c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d37f64",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
