{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2dd7c232",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "og_df = pd.read_pickle('df.pkl') \n",
    "count_df1 = og_df.groupby(['transcript_id', 'transcript_position']).size().reset_index(name='read_count')\n",
    "count_df2 = og_df.groupby(['transcript_id']).size().reset_index(name='expression_count') #most genes are likely to produce unique transcripts\n",
    "df = og_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "df = df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'sum'\n",
    "}\n",
    "\n",
    "summary_df = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "\n",
    "summary_df = summary_df.merge(count_df1, on=['transcript_id', 'transcript_position'])\n",
    "summary_df = summary_df.merge(count_df2, on=['transcript_id'])\n",
    "\n",
    "summary_df['label_percentage'] = summary_df['label'] / summary_df['read_count'] * 100\n",
    "summary_df = summary_df.rename(columns={'label': 'label_sum'})\n",
    "summary_df['label'] = summary_df['label_percentage'].apply(lambda x: 0 if x == 0 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0372e87d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agg_functions = {\n",
    "    'gene_id': 'first',\n",
    "    'combined nucleotides': 'first',\n",
    "    'dwelling_time1': 'mean',\n",
    "    'sd1': 'mean',\n",
    "    'mean1': 'mean',\n",
    "    'dwelling_time2': 'mean',\n",
    "    'sd2': 'mean',\n",
    "    'mean2': 'mean',\n",
    "    'dwelling_time3': 'mean',\n",
    "    'sd3': 'mean',\n",
    "    'mean3': 'mean',\n",
    "    'label': 'first'\n",
    "}\n",
    "\n",
    "summary_df1 = og_df.groupby(['transcript_id', 'transcript_position']).agg(agg_functions).reset_index()\n",
    "summary_df1['label'].equals(summary_df['label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "571e8fb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0., 100.])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df['label_percentage'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3bfee491",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>sd3</th>\n",
       "      <th>mean3</th>\n",
       "      <th>read_count</th>\n",
       "      <th>expression_count</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>185</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>172</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>185</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>200</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>198</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>4.005616</td>\n",
       "      <td>82.004110</td>\n",
       "      <td>73</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>3.644638</td>\n",
       "      <td>80.497101</td>\n",
       "      <td>69</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>2.181563</td>\n",
       "      <td>84.190625</td>\n",
       "      <td>64</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>2.540877</td>\n",
       "      <td>82.289474</td>\n",
       "      <td>57</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>2.576731</td>\n",
       "      <td>78.536538</td>\n",
       "      <td>52</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_position          gene_id  \\\n",
       "0       ENST00000000233                  244  ENSG00000004059   \n",
       "1       ENST00000000233                  261  ENSG00000004059   \n",
       "2       ENST00000000233                  316  ENSG00000004059   \n",
       "3       ENST00000000233                  332  ENSG00000004059   \n",
       "4       ENST00000000233                  368  ENSG00000004059   \n",
       "...                 ...                  ...              ...   \n",
       "121833  ENST00000641834                 1348  ENSG00000167747   \n",
       "121834  ENST00000641834                 1429  ENSG00000167747   \n",
       "121835  ENST00000641834                 1531  ENSG00000167747   \n",
       "121836  ENST00000641834                 1537  ENSG00000167747   \n",
       "121837  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "       combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                   AAGACCA        0.008264  4.223784  123.702703   \n",
       "1                   CAAACTG        0.006609  3.216424  109.681395   \n",
       "2                   GAAACAG        0.007570  2.940541  105.475676   \n",
       "3                   AGAACAT        0.010620  6.476350  129.355000   \n",
       "4                   AGGACAA        0.010701  6.415051  117.924242   \n",
       "...                     ...             ...       ...         ...   \n",
       "121833              GGGACAT        0.009594  3.294164  118.232877   \n",
       "121834              CTGACAC        0.008393  4.511014  110.969565   \n",
       "121835              TGGACAC        0.008161  3.918438  113.968750   \n",
       "121836              CTGACCA        0.008044  3.191228  109.354386   \n",
       "121837              TTGACAT        0.008788  4.090577  105.807692   \n",
       "\n",
       "        dwelling_time2       sd2       mean2  dwelling_time3       sd3  \\\n",
       "0             0.009373  7.382162  125.913514        0.007345  4.386989   \n",
       "1             0.006813  3.226535  107.889535        0.007710  3.016599   \n",
       "2             0.007416  3.642703   98.947027        0.007555  2.087146   \n",
       "3             0.008632  2.899200   97.836500        0.006101  2.236520   \n",
       "4             0.011479  5.870303  121.954545        0.010019  4.260253   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.007300  4.929726  116.342466        0.006555  4.005616   \n",
       "121834        0.010305  9.105797  114.927536        0.005568  3.644638   \n",
       "121835        0.006877  4.759688  113.562500        0.006410  2.181563   \n",
       "121836        0.007419  6.552982  123.263158        0.006472  2.540877   \n",
       "121837        0.006908  8.702885  113.134615        0.008337  2.576731   \n",
       "\n",
       "            mean3  read_count  expression_count  label  \n",
       "0       80.570270         185              3449      0  \n",
       "1       94.290698         172              3449      0  \n",
       "2       89.364324         185              3449      0  \n",
       "3       89.154000         200              3449      0  \n",
       "4       85.178788         198              3449      0  \n",
       "...           ...         ...               ...    ...  \n",
       "121833  82.004110          73              1323      1  \n",
       "121834  80.497101          69              1323      0  \n",
       "121835  84.190625          64              1323      1  \n",
       "121836  82.289474          57              1323      0  \n",
       "121837  78.536538          52              1323      0  \n",
       "\n",
       "[121838 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = summary_df.drop(columns=['label_percentage', 'label_sum'])\n",
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "049e84be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "label\n",
      "0    10527532\n",
      "1      499574\n",
      "Name: count, dtype: int64\n",
      "label\n",
      "0    116363\n",
      "1      5475\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['label'].value_counts())\n",
    "print(summary_df['label'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63d56caa",
   "metadata": {},
   "source": [
    "### 5-mer feature and binary features for first and last letters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "24380451",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to convert features\n",
    "def sequence_to_features(seq):\n",
    "    \"\"\"\n",
    "    Convert a 7-letter sequence into binary features for the first and last letters.\n",
    "    \n",
    "    Parameters:\n",
    "    - seq: A string of 7 letters (A, T, C, G)\n",
    "    \n",
    "    Returns:\n",
    "    - A dictionary with binary features for the first and last letters.\n",
    "    \"\"\"\n",
    "    letters = ['A', 'T', 'C', 'G']\n",
    "    features = {}\n",
    "    \n",
    "    first_char = seq[0]\n",
    "    last_char = seq[-1]\n",
    "    \n",
    "    for letter in letters:\n",
    "        features[f\"1{letter}\"] = first_char == letter\n",
    "        features[f\"7{letter}\"] = last_char == letter\n",
    "            \n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f0d3c294",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = summary_df.copy()\n",
    "\n",
    "feature_dicts = new_df['combined nucleotides'].apply(sequence_to_features)\n",
    "features_df = feature_dicts.apply(pd.Series)\n",
    "new_df = pd.concat([new_df, features_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0a65240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>expression_count</th>\n",
       "      <th>label</th>\n",
       "      <th>1A</th>\n",
       "      <th>7A</th>\n",
       "      <th>1T</th>\n",
       "      <th>7T</th>\n",
       "      <th>1C</th>\n",
       "      <th>7C</th>\n",
       "      <th>1G</th>\n",
       "      <th>7G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>3449</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>1323</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_position          gene_id  \\\n",
       "0       ENST00000000233                  244  ENSG00000004059   \n",
       "1       ENST00000000233                  261  ENSG00000004059   \n",
       "2       ENST00000000233                  316  ENSG00000004059   \n",
       "3       ENST00000000233                  332  ENSG00000004059   \n",
       "4       ENST00000000233                  368  ENSG00000004059   \n",
       "...                 ...                  ...              ...   \n",
       "121833  ENST00000641834                 1348  ENSG00000167747   \n",
       "121834  ENST00000641834                 1429  ENSG00000167747   \n",
       "121835  ENST00000641834                 1531  ENSG00000167747   \n",
       "121836  ENST00000641834                 1537  ENSG00000167747   \n",
       "121837  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "       combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                   AAGACCA        0.008264  4.223784  123.702703   \n",
       "1                   CAAACTG        0.006609  3.216424  109.681395   \n",
       "2                   GAAACAG        0.007570  2.940541  105.475676   \n",
       "3                   AGAACAT        0.010620  6.476350  129.355000   \n",
       "4                   AGGACAA        0.010701  6.415051  117.924242   \n",
       "...                     ...             ...       ...         ...   \n",
       "121833              GGGACAT        0.009594  3.294164  118.232877   \n",
       "121834              CTGACAC        0.008393  4.511014  110.969565   \n",
       "121835              TGGACAC        0.008161  3.918438  113.968750   \n",
       "121836              CTGACCA        0.008044  3.191228  109.354386   \n",
       "121837              TTGACAT        0.008788  4.090577  105.807692   \n",
       "\n",
       "        dwelling_time2       sd2       mean2  ...  expression_count  label  \\\n",
       "0             0.009373  7.382162  125.913514  ...              3449      0   \n",
       "1             0.006813  3.226535  107.889535  ...              3449      0   \n",
       "2             0.007416  3.642703   98.947027  ...              3449      0   \n",
       "3             0.008632  2.899200   97.836500  ...              3449      0   \n",
       "4             0.011479  5.870303  121.954545  ...              3449      0   \n",
       "...                ...       ...         ...  ...               ...    ...   \n",
       "121833        0.007300  4.929726  116.342466  ...              1323      1   \n",
       "121834        0.010305  9.105797  114.927536  ...              1323      0   \n",
       "121835        0.006877  4.759688  113.562500  ...              1323      1   \n",
       "121836        0.007419  6.552982  123.263158  ...              1323      0   \n",
       "121837        0.006908  8.702885  113.134615  ...              1323      0   \n",
       "\n",
       "           1A     7A     1T     7T     1C     7C     1G     7G  \n",
       "0        True   True  False  False  False  False  False  False  \n",
       "1       False  False  False  False   True  False  False   True  \n",
       "2       False  False  False  False  False  False   True   True  \n",
       "3        True  False  False   True  False  False  False  False  \n",
       "4        True   True  False  False  False  False  False  False  \n",
       "...       ...    ...    ...    ...    ...    ...    ...    ...  \n",
       "121833  False  False  False   True  False  False   True  False  \n",
       "121834  False  False  False  False   True   True  False  False  \n",
       "121835  False  False   True  False  False   True  False  False  \n",
       "121836  False   True  False  False   True  False  False  False  \n",
       "121837  False  False   True   True  False  False  False  False  \n",
       "\n",
       "[121838 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "75cdd2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5-mer\n",
       "GGACT    1440\n",
       "GAACT     783\n",
       "GGACA     691\n",
       "GGACC     552\n",
       "AGACT     525\n",
       "TGACT     430\n",
       "AAACT     255\n",
       "GAACA     187\n",
       "AGACA     147\n",
       "GAACC     124\n",
       "TGACC      90\n",
       "TAACT      71\n",
       "TGACA      68\n",
       "AGACC      60\n",
       "AAACA      38\n",
       "AAACC      11\n",
       "TAACC       2\n",
       "TAACA       1\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df['5-mer'] = new_df['combined nucleotides'].apply(lambda x: x[1:-1])\n",
    "new_df.groupby(['5-mer'])['label'].sum().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0ad2245e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>label</th>\n",
       "      <th>1A</th>\n",
       "      <th>7A</th>\n",
       "      <th>1T</th>\n",
       "      <th>7T</th>\n",
       "      <th>1C</th>\n",
       "      <th>7C</th>\n",
       "      <th>1G</th>\n",
       "      <th>7G</th>\n",
       "      <th>5-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_position          gene_id  \\\n",
       "0       ENST00000000233                  244  ENSG00000004059   \n",
       "1       ENST00000000233                  261  ENSG00000004059   \n",
       "2       ENST00000000233                  316  ENSG00000004059   \n",
       "3       ENST00000000233                  332  ENSG00000004059   \n",
       "4       ENST00000000233                  368  ENSG00000004059   \n",
       "...                 ...                  ...              ...   \n",
       "121833  ENST00000641834                 1348  ENSG00000167747   \n",
       "121834  ENST00000641834                 1429  ENSG00000167747   \n",
       "121835  ENST00000641834                 1531  ENSG00000167747   \n",
       "121836  ENST00000641834                 1537  ENSG00000167747   \n",
       "121837  ENST00000641834                 1693  ENSG00000167747   \n",
       "\n",
       "       combined nucleotides  dwelling_time1       sd1       mean1  \\\n",
       "0                   AAGACCA        0.008264  4.223784  123.702703   \n",
       "1                   CAAACTG        0.006609  3.216424  109.681395   \n",
       "2                   GAAACAG        0.007570  2.940541  105.475676   \n",
       "3                   AGAACAT        0.010620  6.476350  129.355000   \n",
       "4                   AGGACAA        0.010701  6.415051  117.924242   \n",
       "...                     ...             ...       ...         ...   \n",
       "121833              GGGACAT        0.009594  3.294164  118.232877   \n",
       "121834              CTGACAC        0.008393  4.511014  110.969565   \n",
       "121835              TGGACAC        0.008161  3.918438  113.968750   \n",
       "121836              CTGACCA        0.008044  3.191228  109.354386   \n",
       "121837              TTGACAT        0.008788  4.090577  105.807692   \n",
       "\n",
       "        dwelling_time2       sd2       mean2  ...  label     1A     7A     1T  \\\n",
       "0             0.009373  7.382162  125.913514  ...      0   True   True  False   \n",
       "1             0.006813  3.226535  107.889535  ...      0  False  False  False   \n",
       "2             0.007416  3.642703   98.947027  ...      0  False  False  False   \n",
       "3             0.008632  2.899200   97.836500  ...      0   True  False  False   \n",
       "4             0.011479  5.870303  121.954545  ...      0   True   True  False   \n",
       "...                ...       ...         ...  ...    ...    ...    ...    ...   \n",
       "121833        0.007300  4.929726  116.342466  ...      1  False  False  False   \n",
       "121834        0.010305  9.105797  114.927536  ...      0  False  False  False   \n",
       "121835        0.006877  4.759688  113.562500  ...      1  False  False   True   \n",
       "121836        0.007419  6.552982  123.263158  ...      0  False   True  False   \n",
       "121837        0.006908  8.702885  113.134615  ...      0  False  False   True   \n",
       "\n",
       "           7T     1C     7C     1G     7G  5-mer  \n",
       "0       False  False  False  False  False  AGACC  \n",
       "1       False   True  False  False   True  AAACT  \n",
       "2       False  False  False   True   True  AAACA  \n",
       "3        True  False  False  False  False  GAACA  \n",
       "4       False  False  False  False  False  GGACA  \n",
       "...       ...    ...    ...    ...    ...    ...  \n",
       "121833   True  False  False   True  False  GGACA  \n",
       "121834  False   True   True  False  False  TGACA  \n",
       "121835  False  False   True  False  False  GGACA  \n",
       "121836  False   True  False  False  False  TGACC  \n",
       "121837   True  False  False  False  False  TGACA  \n",
       "\n",
       "[121838 rows x 25 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c75b25",
   "metadata": {},
   "source": [
    "### Label Encoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4709d4fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>label</th>\n",
       "      <th>1A</th>\n",
       "      <th>7A</th>\n",
       "      <th>1T</th>\n",
       "      <th>7T</th>\n",
       "      <th>1C</th>\n",
       "      <th>7C</th>\n",
       "      <th>1G</th>\n",
       "      <th>7G</th>\n",
       "      <th>5-mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>GGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TGACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                   0                  244       10              AAGACCA   \n",
       "1                   0                  261       10              CAAACTG   \n",
       "2                   0                  316       10              GAAACAG   \n",
       "3                   0                  332       10              AGAACAT   \n",
       "4                   0                  368       10              AGGACAA   \n",
       "...               ...                  ...      ...                  ...   \n",
       "121833           5332                 1348     2764              GGGACAT   \n",
       "121834           5332                 1429     2764              CTGACAC   \n",
       "121835           5332                 1531     2764              TGGACAC   \n",
       "121836           5332                 1537     2764              CTGACCA   \n",
       "121837           5332                 1693     2764              TTGACAT   \n",
       "\n",
       "        dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0             0.008264  4.223784  123.702703        0.009373  7.382162   \n",
       "1             0.006609  3.216424  109.681395        0.006813  3.226535   \n",
       "2             0.007570  2.940541  105.475676        0.007416  3.642703   \n",
       "3             0.010620  6.476350  129.355000        0.008632  2.899200   \n",
       "4             0.010701  6.415051  117.924242        0.011479  5.870303   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.009594  3.294164  118.232877        0.007300  4.929726   \n",
       "121834        0.008393  4.511014  110.969565        0.010305  9.105797   \n",
       "121835        0.008161  3.918438  113.968750        0.006877  4.759688   \n",
       "121836        0.008044  3.191228  109.354386        0.007419  6.552982   \n",
       "121837        0.008788  4.090577  105.807692        0.006908  8.702885   \n",
       "\n",
       "             mean2  ...  label     1A     7A     1T     7T     1C     7C  \\\n",
       "0       125.913514  ...      0   True   True  False  False  False  False   \n",
       "1       107.889535  ...      0  False  False  False  False   True  False   \n",
       "2        98.947027  ...      0  False  False  False  False  False  False   \n",
       "3        97.836500  ...      0   True  False  False   True  False  False   \n",
       "4       121.954545  ...      0   True   True  False  False  False  False   \n",
       "...            ...  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "121833  116.342466  ...      1  False  False  False   True  False  False   \n",
       "121834  114.927536  ...      0  False  False  False  False   True   True   \n",
       "121835  113.562500  ...      1  False  False   True  False  False   True   \n",
       "121836  123.263158  ...      0  False   True  False  False   True  False   \n",
       "121837  113.134615  ...      0  False  False   True   True  False  False   \n",
       "\n",
       "           1G     7G  5-mer  \n",
       "0       False  False  AGACC  \n",
       "1       False   True  AAACT  \n",
       "2        True   True  AAACA  \n",
       "3       False  False  GAACA  \n",
       "4       False  False  GGACA  \n",
       "...       ...    ...    ...  \n",
       "121833   True  False  GGACA  \n",
       "121834  False  False  TGACA  \n",
       "121835  False  False  GGACA  \n",
       "121836  False  False  TGACC  \n",
       "121837  False  False  TGACA  \n",
       "\n",
       "[121838 rows x 25 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "combined = new_df.copy()\n",
    "\n",
    "label_encoder = LabelEncoder()\n",
    "combined['transcript_id'] = label_encoder.fit_transform(combined['transcript_id'])\n",
    "combined['gene_id'] = label_encoder.fit_transform(combined['gene_id'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5e3c240a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>combined nucleotides</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id combined nucleotides  \\\n",
       "0                   0                  244       10              AAGACCA   \n",
       "1                   0                  261       10              CAAACTG   \n",
       "2                   0                  316       10              GAAACAG   \n",
       "3                   0                  332       10              AGAACAT   \n",
       "4                   0                  368       10              AGGACAA   \n",
       "...               ...                  ...      ...                  ...   \n",
       "121833           5332                 1348     2764              GGGACAT   \n",
       "121834           5332                 1429     2764              CTGACAC   \n",
       "121835           5332                 1531     2764              TGGACAC   \n",
       "121836           5332                 1537     2764              CTGACCA   \n",
       "121837           5332                 1693     2764              TTGACAT   \n",
       "\n",
       "        dwelling_time1       sd1       mean1  dwelling_time2       sd2  \\\n",
       "0             0.008264  4.223784  123.702703        0.009373  7.382162   \n",
       "1             0.006609  3.216424  109.681395        0.006813  3.226535   \n",
       "2             0.007570  2.940541  105.475676        0.007416  3.642703   \n",
       "3             0.010620  6.476350  129.355000        0.008632  2.899200   \n",
       "4             0.010701  6.415051  117.924242        0.011479  5.870303   \n",
       "...                ...       ...         ...             ...       ...   \n",
       "121833        0.009594  3.294164  118.232877        0.007300  4.929726   \n",
       "121834        0.008393  4.511014  110.969565        0.010305  9.105797   \n",
       "121835        0.008161  3.918438  113.968750        0.006877  4.759688   \n",
       "121836        0.008044  3.191228  109.354386        0.007419  6.552982   \n",
       "121837        0.008788  4.090577  105.807692        0.006908  8.702885   \n",
       "\n",
       "             mean2  ...  5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  \\\n",
       "0       125.913514  ...        False        False        False        False   \n",
       "1       107.889535  ...        False        False        False        False   \n",
       "2        98.947027  ...        False        False        False        False   \n",
       "3        97.836500  ...        False        False        False        False   \n",
       "4       121.954545  ...        False         True        False        False   \n",
       "...            ...  ...          ...          ...          ...          ...   \n",
       "121833  116.342466  ...        False         True        False        False   \n",
       "121834  114.927536  ...        False        False        False        False   \n",
       "121835  113.562500  ...        False         True        False        False   \n",
       "121836  123.263158  ...        False        False        False        False   \n",
       "121837  113.134615  ...        False        False        False        False   \n",
       "\n",
       "        5-mer_TAACA  5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  \\\n",
       "0             False        False        False        False        False   \n",
       "1             False        False        False        False        False   \n",
       "2             False        False        False        False        False   \n",
       "3             False        False        False        False        False   \n",
       "4             False        False        False        False        False   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "121833        False        False        False        False        False   \n",
       "121834        False        False        False         True        False   \n",
       "121835        False        False        False        False        False   \n",
       "121836        False        False        False        False         True   \n",
       "121837        False        False        False         True        False   \n",
       "\n",
       "        5-mer_TGACT  \n",
       "0             False  \n",
       "1             False  \n",
       "2             False  \n",
       "3             False  \n",
       "4             False  \n",
       "...             ...  \n",
       "121833        False  \n",
       "121834        False  \n",
       "121835        False  \n",
       "121836        False  \n",
       "121837        False  \n",
       "\n",
       "[121838 rows x 42 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = pd.get_dummies(combined, columns=['5-mer'], drop_first=False)\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2969aec6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id  dwelling_time1       sd1  \\\n",
       "0                   0                  244       10        0.008264  4.223784   \n",
       "1                   0                  261       10        0.006609  3.216424   \n",
       "2                   0                  316       10        0.007570  2.940541   \n",
       "3                   0                  332       10        0.010620  6.476350   \n",
       "4                   0                  368       10        0.010701  6.415051   \n",
       "...               ...                  ...      ...             ...       ...   \n",
       "121833           5332                 1348     2764        0.009594  3.294164   \n",
       "121834           5332                 1429     2764        0.008393  4.511014   \n",
       "121835           5332                 1531     2764        0.008161  3.918438   \n",
       "121836           5332                 1537     2764        0.008044  3.191228   \n",
       "121837           5332                 1693     2764        0.008788  4.090577   \n",
       "\n",
       "             mean1  dwelling_time2       sd2       mean2  dwelling_time3  ...  \\\n",
       "0       123.702703        0.009373  7.382162  125.913514        0.007345  ...   \n",
       "1       109.681395        0.006813  3.226535  107.889535        0.007710  ...   \n",
       "2       105.475676        0.007416  3.642703   98.947027        0.007555  ...   \n",
       "3       129.355000        0.008632  2.899200   97.836500        0.006101  ...   \n",
       "4       117.924242        0.011479  5.870303  121.954545        0.010019  ...   \n",
       "...            ...             ...       ...         ...             ...  ...   \n",
       "121833  118.232877        0.007300  4.929726  116.342466        0.006555  ...   \n",
       "121834  110.969565        0.010305  9.105797  114.927536        0.005568  ...   \n",
       "121835  113.968750        0.006877  4.759688  113.562500        0.006410  ...   \n",
       "121836  109.354386        0.007419  6.552982  123.263158        0.006472  ...   \n",
       "121837  105.807692        0.006908  8.702885  113.134615        0.008337  ...   \n",
       "\n",
       "        5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  5-mer_TAACA  \\\n",
       "0             False        False        False        False        False   \n",
       "1             False        False        False        False        False   \n",
       "2             False        False        False        False        False   \n",
       "3             False        False        False        False        False   \n",
       "4             False         True        False        False        False   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "121833        False         True        False        False        False   \n",
       "121834        False        False        False        False        False   \n",
       "121835        False         True        False        False        False   \n",
       "121836        False        False        False        False        False   \n",
       "121837        False        False        False        False        False   \n",
       "\n",
       "        5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  5-mer_TGACT  \n",
       "0             False        False        False        False        False  \n",
       "1             False        False        False        False        False  \n",
       "2             False        False        False        False        False  \n",
       "3             False        False        False        False        False  \n",
       "4             False        False        False        False        False  \n",
       "...             ...          ...          ...          ...          ...  \n",
       "121833        False        False        False        False        False  \n",
       "121834        False        False         True        False        False  \n",
       "121835        False        False        False        False        False  \n",
       "121836        False        False        False         True        False  \n",
       "121837        False        False         True        False        False  \n",
       "\n",
       "[121838 rows x 41 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = combined.drop(columns=['combined nucleotides'])\n",
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2014fac5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "1528    274\n",
       "977     253\n",
       "2632    250\n",
       "2005    230\n",
       "1929    229\n",
       "       ... \n",
       "857       1\n",
       "863       1\n",
       "887       1\n",
       "2800      1\n",
       "1412      1\n",
       "Name: count, Length: 3852, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined.value_counts('gene_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf19172",
   "metadata": {},
   "source": [
    "### Resampling Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e68195be",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = combined.drop('label', axis=1)\n",
    "y = combined['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "00fb4e4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>dwelling_time1</th>\n",
       "      <th>sd1</th>\n",
       "      <th>mean1</th>\n",
       "      <th>dwelling_time2</th>\n",
       "      <th>sd2</th>\n",
       "      <th>mean2</th>\n",
       "      <th>dwelling_time3</th>\n",
       "      <th>...</th>\n",
       "      <th>5-mer_GAACT</th>\n",
       "      <th>5-mer_GGACA</th>\n",
       "      <th>5-mer_GGACC</th>\n",
       "      <th>5-mer_GGACT</th>\n",
       "      <th>5-mer_TAACA</th>\n",
       "      <th>5-mer_TAACC</th>\n",
       "      <th>5-mer_TAACT</th>\n",
       "      <th>5-mer_TGACA</th>\n",
       "      <th>5-mer_TGACC</th>\n",
       "      <th>5-mer_TGACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>10</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>10</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>368</td>\n",
       "      <td>10</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>5332</td>\n",
       "      <td>1348</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>5332</td>\n",
       "      <td>1429</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>5332</td>\n",
       "      <td>1531</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759688</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>5332</td>\n",
       "      <td>1537</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>5332</td>\n",
       "      <td>1693</td>\n",
       "      <td>2764</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  transcript_position  gene_id  dwelling_time1       sd1  \\\n",
       "0                   0                  244       10        0.008264  4.223784   \n",
       "1                   0                  261       10        0.006609  3.216424   \n",
       "2                   0                  316       10        0.007570  2.940541   \n",
       "3                   0                  332       10        0.010620  6.476350   \n",
       "4                   0                  368       10        0.010701  6.415051   \n",
       "...               ...                  ...      ...             ...       ...   \n",
       "121833           5332                 1348     2764        0.009594  3.294164   \n",
       "121834           5332                 1429     2764        0.008393  4.511014   \n",
       "121835           5332                 1531     2764        0.008161  3.918438   \n",
       "121836           5332                 1537     2764        0.008044  3.191228   \n",
       "121837           5332                 1693     2764        0.008788  4.090577   \n",
       "\n",
       "             mean1  dwelling_time2       sd2       mean2  dwelling_time3  ...  \\\n",
       "0       123.702703        0.009373  7.382162  125.913514        0.007345  ...   \n",
       "1       109.681395        0.006813  3.226535  107.889535        0.007710  ...   \n",
       "2       105.475676        0.007416  3.642703   98.947027        0.007555  ...   \n",
       "3       129.355000        0.008632  2.899200   97.836500        0.006101  ...   \n",
       "4       117.924242        0.011479  5.870303  121.954545        0.010019  ...   \n",
       "...            ...             ...       ...         ...             ...  ...   \n",
       "121833  118.232877        0.007300  4.929726  116.342466        0.006555  ...   \n",
       "121834  110.969565        0.010305  9.105797  114.927536        0.005568  ...   \n",
       "121835  113.968750        0.006877  4.759688  113.562500        0.006410  ...   \n",
       "121836  109.354386        0.007419  6.552982  123.263158        0.006472  ...   \n",
       "121837  105.807692        0.006908  8.702885  113.134615        0.008337  ...   \n",
       "\n",
       "        5-mer_GAACT  5-mer_GGACA  5-mer_GGACC  5-mer_GGACT  5-mer_TAACA  \\\n",
       "0             False        False        False        False        False   \n",
       "1             False        False        False        False        False   \n",
       "2             False        False        False        False        False   \n",
       "3             False        False        False        False        False   \n",
       "4             False         True        False        False        False   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "121833        False         True        False        False        False   \n",
       "121834        False        False        False        False        False   \n",
       "121835        False         True        False        False        False   \n",
       "121836        False        False        False        False        False   \n",
       "121837        False        False        False        False        False   \n",
       "\n",
       "        5-mer_TAACC  5-mer_TAACT  5-mer_TGACA  5-mer_TGACC  5-mer_TGACT  \n",
       "0             False        False        False        False        False  \n",
       "1             False        False        False        False        False  \n",
       "2             False        False        False        False        False  \n",
       "3             False        False        False        False        False  \n",
       "4             False        False        False        False        False  \n",
       "...             ...          ...          ...          ...          ...  \n",
       "121833        False        False        False        False        False  \n",
       "121834        False        False         True        False        False  \n",
       "121835        False        False        False        False        False  \n",
       "121836        False        False        False         True        False  \n",
       "121837        False        False         True        False        False  \n",
       "\n",
       "[121838 rows x 40 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "18d86d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.under_sampling import TomekLinks, RandomUnderSampler\n",
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.over_sampling import SMOTE, ADASYN\n",
    "from imblearn.combine import SMOTETomek\n",
    "\n",
    "# ADASYN followed by Tomek Links Pipeline\n",
    "adasyn_tomek = Pipeline([\n",
    "    ('adasyn', ADASYN(random_state=42)),\n",
    "    ('tomek', TomekLinks())\n",
    "])\n",
    "\n",
    "# Dictionary of resampling methods\n",
    "resampling_methods = {\n",
    "    'RandomUnderSampler': RandomUnderSampler(random_state=42),\n",
    "    'TomekLinks': TomekLinks(),\n",
    "    'SMOTE': SMOTE(random_state=42),\n",
    "    'ADASYN': ADASYN(random_state=42),\n",
    "    'SMOTETomek': SMOTETomek(random_state=42),\n",
    "    'ADASYN_Tomek': adasyn_tomek\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f560412",
   "metadata": {},
   "source": [
    "### XGBoost with GroupShuffleSplit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a90fdf20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for RandomUnderSampler:\n",
      "ROC AUC scores for each fold: [0.8906834928744976, 0.9025330044261164, 0.8913792961119329, 0.8907848352281983, 0.8908846438822484]\n",
      "Mean ROC AUC score: 0.8932530545045987\n",
      "Standard deviation of ROC AUC scores: 0.004646128782186436\n",
      "--------------------------------------------------\n",
      "Results for TomekLinks:\n",
      "ROC AUC scores for each fold: [0.8947092831417606, 0.9056361937687386, 0.8998074033799295, 0.8962331317024385, 0.8949339361028431]\n",
      "Mean ROC AUC score: 0.898263989619142\n",
      "Standard deviation of ROC AUC scores: 0.0041129332233798095\n",
      "--------------------------------------------------\n",
      "Results for SMOTE:\n",
      "ROC AUC scores for each fold: [0.8879156070698921, 0.8972632747747272, 0.8877335530685254, 0.8815078505606367, 0.8838169433410119]\n",
      "Mean ROC AUC score: 0.8876474457629587\n",
      "Standard deviation of ROC AUC scores: 0.005383509461143215\n",
      "--------------------------------------------------\n",
      "Results for ADASYN:\n",
      "ROC AUC scores for each fold: [0.8859953738233554, 0.8955102943441164, 0.8867371744215629, 0.8778843475280322, 0.882340636549126]\n",
      "Mean ROC AUC score: 0.8856935653332385\n",
      "Standard deviation of ROC AUC scores: 0.005827081446615161\n",
      "--------------------------------------------------\n",
      "Results for SMOTETomek:\n",
      "ROC AUC scores for each fold: [0.8866701582145551, 0.8968079877160446, 0.8889800465361313, 0.8812468482988511, 0.883983785335699]\n",
      "Mean ROC AUC score: 0.8875377652202563\n",
      "Standard deviation of ROC AUC scores: 0.005309921866915631\n",
      "--------------------------------------------------\n",
      "Results for ADASYN_Tomek:\n",
      "ROC AUC scores for each fold: [0.8864883212645709, 0.897282840104772, 0.8870776495238697, 0.87944577595549, 0.8815513426986159]\n",
      "Mean ROC AUC score: 0.8863691859094637\n",
      "Standard deviation of ROC AUC scores: 0.006176879154000639\n",
      "--------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import xgboost as xgb\n",
    "\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, precision_recall_curve\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Model initialization\n",
    "xgb_model = xgb.XGBClassifier(n_estimators=300, random_state=42, learning_rate=0.1, max_depth=5)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X_no_gene_id = X.drop(columns=['gene_id'])\n",
    "\n",
    "# Loop through each resampling method\n",
    "for method_name, resampler in resampling_methods.items():\n",
    "    print(f\"Results for {method_name}:\")\n",
    "    \n",
    "    roc_aucs = []  # List to store ROC AUC scores for each fold\n",
    "    \n",
    "    # Loop through each fold\n",
    "    for train_index, valid_index in group_shuffle_split.split(X, y, groups=X['gene_id']):\n",
    "        X_train_fold, X_valid_fold = X_no_gene_id.iloc[train_index], X_no_gene_id.iloc[valid_index]\n",
    "        y_train_fold, y_valid_fold = y.iloc[train_index], y.iloc[valid_index]\n",
    "        \n",
    "        # Initialize the scaler\n",
    "        scaler = MinMaxScaler()\n",
    "        \n",
    "        # Fit the scaler on the training data and transform both training and validation data\n",
    "        X_train_fold_scaled = scaler.fit_transform(X_train_fold)\n",
    "        X_valid_fold_scaled = scaler.transform(X_valid_fold)\n",
    "        \n",
    "        # Apply the current resampling technique\n",
    "        X_resampled, y_resampled = resampler.fit_resample(X_train_fold_scaled, y_train_fold)\n",
    "        \n",
    "        # Train the model on the resampled data\n",
    "        xgb_model.fit(X_resampled, y_resampled)\n",
    "        \n",
    "        # Predict on the validation set\n",
    "        y_pred = xgb_model.predict_proba(X_valid_fold_scaled)[:, 1]\n",
    "        \n",
    "        threshold = 0.4\n",
    "        y_pred_binary = (y_pred >= threshold).astype(int)\n",
    "        \n",
    "        # Compute the ROC AUC score for this fold and append to the list\n",
    "        roc_aucs.append(roc_auc_score(y_valid_fold, y_pred))\n",
    "    \n",
    "    # Display results for the current resampling method\n",
    "    print(\"ROC AUC scores for each fold:\", roc_aucs)\n",
    "    print(\"Mean ROC AUC score:\", np.mean(roc_aucs))\n",
    "    print(\"Standard deviation of ROC AUC scores:\", np.std(roc_aucs))\n",
    "    print(\"-\" * 50)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1610a7",
   "metadata": {},
   "source": [
    "### Gradient Boosting with GroupShuffleSplit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f6949619",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for RandomUnderSampler:\n",
      "ROC AUC scores for each fold: [0.8877166902751392, 0.9030808519527316, 0.8922574770377475, 0.8907830573155072, 0.8920635098650983]\n",
      "Mean ROC AUC score: 0.8931803172892447\n",
      "Standard deviation of ROC AUC scores: 0.00520994347582129\n",
      "--------------------------------------------------\n",
      "Results for TomekLinks:\n",
      "ROC AUC scores for each fold: [0.893745875478534, 0.8985649177833748, 0.8968498337352057, 0.8974150459510892, 0.890453931347678]\n",
      "Mean ROC AUC score: 0.8954059208591764\n",
      "Standard deviation of ROC AUC scores: 0.0029454602487011634\n",
      "--------------------------------------------------\n",
      "Results for SMOTE:\n",
      "ROC AUC scores for each fold: [0.8868696716851079, 0.890790899027647, 0.8815966992442101, 0.8788857334577624, 0.8802821024110032]\n",
      "Mean ROC AUC score: 0.8836850211651461\n",
      "Standard deviation of ROC AUC scores: 0.004463397946006506\n",
      "--------------------------------------------------\n",
      "Results for ADASYN:\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 34\u001b[0m\n\u001b[1;32m     31\u001b[0m X_resampled, y_resampled \u001b[38;5;241m=\u001b[39m resampler\u001b[38;5;241m.\u001b[39mfit_resample(X_train_fold_scaled, y_train_fold)\n\u001b[1;32m     33\u001b[0m \u001b[38;5;66;03m# Train the model on the resampled data\u001b[39;00m\n\u001b[0;32m---> 34\u001b[0m gb_model\u001b[38;5;241m.\u001b[39mfit(X_resampled, y_resampled)\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# Predict on the validation set\u001b[39;00m\n\u001b[1;32m     37\u001b[0m y_pred \u001b[38;5;241m=\u001b[39m gb_model\u001b[38;5;241m.\u001b[39mpredict_proba(X_valid_fold_scaled)[:, \u001b[38;5;241m1\u001b[39m]\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_gb.py:538\u001b[0m, in \u001b[0;36mBaseGradientBoosting.fit\u001b[0;34m(self, X, y, sample_weight, monitor)\u001b[0m\n\u001b[1;32m    535\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_resize_state()\n\u001b[1;32m    537\u001b[0m \u001b[38;5;66;03m# fit the boosting stages\u001b[39;00m\n\u001b[0;32m--> 538\u001b[0m n_stages \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_stages(\n\u001b[1;32m    539\u001b[0m     X,\n\u001b[1;32m    540\u001b[0m     y,\n\u001b[1;32m    541\u001b[0m     raw_predictions,\n\u001b[1;32m    542\u001b[0m     sample_weight,\n\u001b[1;32m    543\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_rng,\n\u001b[1;32m    544\u001b[0m     X_val,\n\u001b[1;32m    545\u001b[0m     y_val,\n\u001b[1;32m    546\u001b[0m     sample_weight_val,\n\u001b[1;32m    547\u001b[0m     begin_at_stage,\n\u001b[1;32m    548\u001b[0m     monitor,\n\u001b[1;32m    549\u001b[0m )\n\u001b[1;32m    551\u001b[0m \u001b[38;5;66;03m# change shape of arrays after fit (early-stopping or additional ests)\u001b[39;00m\n\u001b[1;32m    552\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m n_stages \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mestimators_\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_gb.py:615\u001b[0m, in \u001b[0;36mBaseGradientBoosting._fit_stages\u001b[0;34m(self, X, y, raw_predictions, sample_weight, random_state, X_val, y_val, sample_weight_val, begin_at_stage, monitor)\u001b[0m\n\u001b[1;32m    608\u001b[0m     old_oob_score \u001b[38;5;241m=\u001b[39m loss_(\n\u001b[1;32m    609\u001b[0m         y[\u001b[38;5;241m~\u001b[39msample_mask],\n\u001b[1;32m    610\u001b[0m         raw_predictions[\u001b[38;5;241m~\u001b[39msample_mask],\n\u001b[1;32m    611\u001b[0m         sample_weight[\u001b[38;5;241m~\u001b[39msample_mask],\n\u001b[1;32m    612\u001b[0m     )\n\u001b[1;32m    614\u001b[0m \u001b[38;5;66;03m# fit next stage of trees\u001b[39;00m\n\u001b[0;32m--> 615\u001b[0m raw_predictions \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_stage(\n\u001b[1;32m    616\u001b[0m     i,\n\u001b[1;32m    617\u001b[0m     X,\n\u001b[1;32m    618\u001b[0m     y,\n\u001b[1;32m    619\u001b[0m     raw_predictions,\n\u001b[1;32m    620\u001b[0m     sample_weight,\n\u001b[1;32m    621\u001b[0m     sample_mask,\n\u001b[1;32m    622\u001b[0m     random_state,\n\u001b[1;32m    623\u001b[0m     X_csc,\n\u001b[1;32m    624\u001b[0m     X_csr,\n\u001b[1;32m    625\u001b[0m )\n\u001b[1;32m    627\u001b[0m \u001b[38;5;66;03m# track deviance (= loss)\u001b[39;00m\n\u001b[1;32m    628\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m do_oob:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_gb.py:257\u001b[0m, in \u001b[0;36mBaseGradientBoosting._fit_stage\u001b[0;34m(self, i, X, y, raw_predictions, sample_weight, sample_mask, random_state, X_csc, X_csr)\u001b[0m\n\u001b[1;32m    254\u001b[0m     sample_weight \u001b[38;5;241m=\u001b[39m sample_weight \u001b[38;5;241m*\u001b[39m sample_mask\u001b[38;5;241m.\u001b[39mastype(np\u001b[38;5;241m.\u001b[39mfloat64)\n\u001b[1;32m    256\u001b[0m X \u001b[38;5;241m=\u001b[39m X_csr \u001b[38;5;28;01mif\u001b[39;00m X_csr \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m X\n\u001b[0;32m--> 257\u001b[0m tree\u001b[38;5;241m.\u001b[39mfit(X, residual, sample_weight\u001b[38;5;241m=\u001b[39msample_weight, check_input\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    259\u001b[0m \u001b[38;5;66;03m# update tree leaves\u001b[39;00m\n\u001b[1;32m    260\u001b[0m loss\u001b[38;5;241m.\u001b[39mupdate_terminal_regions(\n\u001b[1;32m    261\u001b[0m     tree\u001b[38;5;241m.\u001b[39mtree_,\n\u001b[1;32m    262\u001b[0m     X,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    269\u001b[0m     k\u001b[38;5;241m=\u001b[39mk,\n\u001b[1;32m    270\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/sklearn/tree/_classes.py:1247\u001b[0m, in \u001b[0;36mDecisionTreeRegressor.fit\u001b[0;34m(self, X, y, sample_weight, check_input)\u001b[0m\n\u001b[1;32m   1218\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfit\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, y, sample_weight\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, check_input\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[1;32m   1219\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Build a decision tree regressor from the training set (X, y).\u001b[39;00m\n\u001b[1;32m   1220\u001b[0m \n\u001b[1;32m   1221\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1244\u001b[0m \u001b[38;5;124;03m        Fitted estimator.\u001b[39;00m\n\u001b[1;32m   1245\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1247\u001b[0m     \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mfit(\n\u001b[1;32m   1248\u001b[0m         X,\n\u001b[1;32m   1249\u001b[0m         y,\n\u001b[1;32m   1250\u001b[0m         sample_weight\u001b[38;5;241m=\u001b[39msample_weight,\n\u001b[1;32m   1251\u001b[0m         check_input\u001b[38;5;241m=\u001b[39mcheck_input,\n\u001b[1;32m   1252\u001b[0m     )\n\u001b[1;32m   1253\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/sklearn/tree/_classes.py:379\u001b[0m, in \u001b[0;36mBaseDecisionTree.fit\u001b[0;34m(self, X, y, sample_weight, check_input)\u001b[0m\n\u001b[1;32m    368\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    369\u001b[0m     builder \u001b[38;5;241m=\u001b[39m BestFirstTreeBuilder(\n\u001b[1;32m    370\u001b[0m         splitter,\n\u001b[1;32m    371\u001b[0m         min_samples_split,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    376\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmin_impurity_decrease,\n\u001b[1;32m    377\u001b[0m     )\n\u001b[0;32m--> 379\u001b[0m builder\u001b[38;5;241m.\u001b[39mbuild(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtree_, X, y, sample_weight)\n\u001b[1;32m    381\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_outputs_ \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m is_classifier(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    382\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_classes_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_classes_[\u001b[38;5;241m0\u001b[39m]\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "\n",
    "# Model initialization using Gradient Boosting\n",
    "gb_model = GradientBoostingClassifier(n_estimators=200, random_state=42, learning_rate=0.1, max_depth=5)\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X_no_gene_id = X.drop(columns=['gene_id'])\n",
    "\n",
    "# Loop through each resampling method\n",
    "for method_name, resampler in resampling_methods.items():\n",
    "    print(f\"Results for {method_name}:\")\n",
    "    \n",
    "    roc_aucs = []  # List to store ROC AUC scores for each fold\n",
    "    \n",
    "    # Loop through each fold\n",
    "    for train_index, valid_index in group_shuffle_split.split(X, y, groups=X['gene_id']):\n",
    "        X_train_fold, X_valid_fold = X_no_gene_id.iloc[train_index], X_no_gene_id.iloc[valid_index]\n",
    "        y_train_fold, y_valid_fold = y.iloc[train_index], y.iloc[valid_index]\n",
    "        \n",
    "        # Initialize the scaler\n",
    "        scaler = MinMaxScaler()\n",
    "        \n",
    "        # Fit the scaler on the training data and transform both training and validation data\n",
    "        X_train_fold_scaled = scaler.fit_transform(X_train_fold)\n",
    "        X_valid_fold_scaled = scaler.transform(X_valid_fold)\n",
    "        \n",
    "        # Apply the current resampling technique\n",
    "        X_resampled, y_resampled = resampler.fit_resample(X_train_fold_scaled, y_train_fold)\n",
    "        \n",
    "        # Train the model on the resampled data\n",
    "        gb_model.fit(X_resampled, y_resampled)\n",
    "        \n",
    "        # Predict on the validation set\n",
    "        y_pred = gb_model.predict_proba(X_valid_fold_scaled)[:, 1]\n",
    "        \n",
    "        threshold = 0.4\n",
    "        y_pred_binary = (y_pred >= threshold).astype(int)\n",
    "        \n",
    "        # Compute the ROC AUC score for this fold and append to the list\n",
    "        roc_aucs.append(roc_auc_score(y_valid_fold, y_pred))\n",
    "    \n",
    "    print(\"ROC AUC scores for each fold:\", roc_aucs)\n",
    "    print(\"Mean ROC AUC score:\", np.mean(roc_aucs))\n",
    "    print(\"Standard deviation of ROC AUC scores:\", np.std(roc_aucs))\n",
    "    print(\"-\" * 50)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf1e392",
   "metadata": {},
   "source": [
    "### 5-mer only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "df1dbb35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROC AUC scores for each fold: [0.8817122298251977, 0.8986595262110875, 0.8834479006959032, 0.8850225266216694, 0.8837309428909161]\n",
      "Mean ROC AUC score: 0.8865146252489549\n",
      "Standard deviation of ROC AUC scores: 0.006163467635259012\n",
      "--------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import xgboost as xgb\n",
    "\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, precision_recall_curve\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Model initialization\n",
    "xgb_model = xgb.XGBClassifier(n_estimators=300, random_state=42, learning_rate=0.1, max_depth=5)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X_min_features = X.drop(columns=['gene_id', 'transcript_id', 'transcript_position', '1A', '1C', '1G', '1T',\n",
    "                                '7A', '7C', '7G', '7T'])\n",
    "\n",
    "    \n",
    "roc_aucs = []  # List to store ROC AUC scores for each fold\n",
    "    \n",
    "# Loop through each fold\n",
    "for train_index, valid_index in group_shuffle_split.split(X, y, groups=X['gene_id']):\n",
    "    X_train_fold, X_valid_fold = X_min_features.iloc[train_index], X_min_features.iloc[valid_index]\n",
    "    y_train_fold, y_valid_fold = y.iloc[train_index], y.iloc[valid_index]\n",
    "        \n",
    "    # Initialize the scaler\n",
    "    scaler = MinMaxScaler()\n",
    "        \n",
    "    # Fit the scaler on the training data and transform both training and validation data\n",
    "    X_train_fold_scaled = scaler.fit_transform(X_train_fold)\n",
    "    X_valid_fold_scaled = scaler.transform(X_valid_fold)\n",
    "        \n",
    "    # Apply the current resampling technique\n",
    "    X_resampled, y_resampled = TomekLinks().fit_resample(X_train_fold_scaled, y_train_fold)\n",
    "        \n",
    "    # Train the model on the resampled data\n",
    "    xgb_model.fit(X_resampled, y_resampled)\n",
    "        \n",
    "    # Predict on the validation set\n",
    "    y_pred = xgb_model.predict_proba(X_valid_fold_scaled)[:, 1]\n",
    "        \n",
    "    threshold = 0.4\n",
    "    y_pred_binary = (y_pred >= threshold).astype(int)\n",
    "        \n",
    "    # Compute the ROC AUC score for this fold and append to the list\n",
    "    roc_aucs.append(roc_auc_score(y_valid_fold, y_pred))\n",
    "    \n",
    "# Display results for the current resampling method\n",
    "print(\"ROC AUC scores for each fold:\", roc_aucs)\n",
    "print(\"Mean ROC AUC score:\", np.mean(roc_aucs))\n",
    "print(\"Standard deviation of ROC AUC scores:\", np.std(roc_aucs))\n",
    "print(\"-\" * 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ca7b63",
   "metadata": {},
   "source": [
    "### 5-mer & +1 -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3661850d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROC AUC scores for each fold: [0.8968412803588648, 0.9053943150903879, 0.8984643114175997, 0.8974740679737195, 0.8941344365914438]\n",
      "Mean ROC AUC score: 0.8984616822864032\n",
      "Standard deviation of ROC AUC scores: 0.0037520006462992187\n",
      "--------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import xgboost as xgb\n",
    "\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, precision_recall_curve\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Model initialization\n",
    "xgb_model = xgb.XGBClassifier(n_estimators=300, random_state=42, learning_rate=0.1, max_depth=5)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X_min_features = X.drop(columns=['gene_id', 'transcript_id', 'transcript_position'])\n",
    "\n",
    "    \n",
    "roc_aucs = []  # List to store ROC AUC scores for each fold\n",
    "    \n",
    "# Loop through each fold\n",
    "for train_index, valid_index in group_shuffle_split.split(X, y, groups=X['gene_id']):\n",
    "    X_train_fold, X_valid_fold = X_min_features.iloc[train_index], X_min_features.iloc[valid_index]\n",
    "    y_train_fold, y_valid_fold = y.iloc[train_index], y.iloc[valid_index]\n",
    "        \n",
    "    # Initialize the scaler\n",
    "    scaler = MinMaxScaler()\n",
    "        \n",
    "    # Fit the scaler on the training data and transform both training and validation data\n",
    "    X_train_fold_scaled = scaler.fit_transform(X_train_fold)\n",
    "    X_valid_fold_scaled = scaler.transform(X_valid_fold)\n",
    "        \n",
    "    # Apply the current resampling technique\n",
    "    X_resampled, y_resampled = TomekLinks().fit_resample(X_train_fold_scaled, y_train_fold)\n",
    "        \n",
    "    # Train the model on the resampled data\n",
    "    xgb_model.fit(X_resampled, y_resampled)\n",
    "        \n",
    "    # Predict on the validation set\n",
    "    y_pred = xgb_model.predict_proba(X_valid_fold_scaled)[:, 1]\n",
    "        \n",
    "    threshold = 0.4\n",
    "    y_pred_binary = (y_pred >= threshold).astype(int)\n",
    "        \n",
    "    # Compute the ROC AUC score for this fold and append to the list\n",
    "    roc_aucs.append(roc_auc_score(y_valid_fold, y_pred))\n",
    "    \n",
    "# Display results for the current resampling method\n",
    "print(\"ROC AUC scores for each fold:\", roc_aucs)\n",
    "print(\"Mean ROC AUC score:\", np.mean(roc_aucs))\n",
    "print(\"Standard deviation of ROC AUC scores:\", np.std(roc_aucs))\n",
    "print(\"-\" * 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b8e3be",
   "metadata": {},
   "source": [
    "### 5-mer & +1 -1 & transcript_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d38eef8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROC AUC scores for each fold: [0.8956453928871472, 0.9047844070682846, 0.9008629746951812, 0.8988256559164398, 0.8947647950307028]\n",
      "Mean ROC AUC score: 0.8989766451195511\n",
      "Standard deviation of ROC AUC scores: 0.0036372533171979233\n",
      "--------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import xgboost as xgb\n",
    "\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score, roc_curve, precision_recall_curve\n",
    "\n",
    "# Initialize GroupShuffleSplit\n",
    "group_shuffle_split = GroupShuffleSplit(n_splits=5, test_size=0.2, random_state=42)\n",
    "\n",
    "# Model initialization\n",
    "xgb_model = xgb.XGBClassifier(n_estimators=300, random_state=42, learning_rate=0.1, max_depth=5)\n",
    "\n",
    "# Drop 'gene_id' from X\n",
    "X_min_features = X.drop(columns=['gene_id', 'transcript_id'])\n",
    "\n",
    "    \n",
    "roc_aucs = []  # List to store ROC AUC scores for each fold\n",
    "    \n",
    "# Loop through each fold\n",
    "for train_index, valid_index in group_shuffle_split.split(X, y, groups=X['gene_id']):\n",
    "    X_train_fold, X_valid_fold = X_min_features.iloc[train_index], X_min_features.iloc[valid_index]\n",
    "    y_train_fold, y_valid_fold = y.iloc[train_index], y.iloc[valid_index]\n",
    "        \n",
    "    # Initialize the scaler\n",
    "    scaler = MinMaxScaler()\n",
    "        \n",
    "    # Fit the scaler on the training data and transform both training and validation data\n",
    "    X_train_fold_scaled = scaler.fit_transform(X_train_fold)\n",
    "    X_valid_fold_scaled = scaler.transform(X_valid_fold)\n",
    "        \n",
    "    # Apply the current resampling technique\n",
    "    X_resampled, y_resampled = TomekLinks().fit_resample(X_train_fold_scaled, y_train_fold)\n",
    "        \n",
    "    # Train the model on the resampled data\n",
    "    xgb_model.fit(X_resampled, y_resampled)\n",
    "        \n",
    "    # Predict on the validation set\n",
    "    y_pred = xgb_model.predict_proba(X_valid_fold_scaled)[:, 1]\n",
    "        \n",
    "    threshold = 0.4\n",
    "    y_pred_binary = (y_pred >= threshold).astype(int)\n",
    "        \n",
    "    # Compute the ROC AUC score for this fold and append to the list\n",
    "    roc_aucs.append(roc_auc_score(y_valid_fold, y_pred))\n",
    "    \n",
    "# Display results for the current resampling method\n",
    "print(\"ROC AUC scores for each fold:\", roc_aucs)\n",
    "print(\"Mean ROC AUC score:\", np.mean(roc_aucs))\n",
    "print(\"Standard deviation of ROC AUC scores:\", np.std(roc_aucs))\n",
    "print(\"-\" * 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef922f13",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
